
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>2.2. Pandas &#8212; A Hands-On Guide to Neural Data Science</title>
    
  <link rel="stylesheet" href="../_static/css/index.73d71520a4ca3b99cfee5594769eaaae.css">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.2d2078699c18a0efb88233928e1cf6ed.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.3da636dd464baa7582d2.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.be0a4a0c39cd630af62a2fcf693f3f06.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="2.3. SciPy" href="SciPy.html" />
    <link rel="prev" title="2.1. NumPy" href="NumPy.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../index.html">
  
  <img src="../_static/brainz.jpeg" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">A Hands-On Guide to Neural Data Science</h1>
  
</a>
</div>

<form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>

<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
  <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../landing_page.html">
   A Hands-On Guide to Neural Data Science
  </a>
 </li>
</ul>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../Chapter_1/Neural_Data_Science.html">
   1. Neural Data Science
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="reference internal" href="Data_Science_In_Python.html">
   2. Data Science in Python
  </a>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="NumPy.html">
     2.1. NumPy
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     2.2. Pandas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="SciPy.html">
     2.3. SciPy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="VersionControl.html">
     2.4. Version Control
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Chapter_3/Statistical_Concepts.html">
   3. Statistical Concepts
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Chapter_4/Neurodata_Without_Borders.html">
   4. Neurodata Without Borders
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Chapter_5/Single-Cell_Electrophysiology.html">
   5. Single-Cell Electrophysiology
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Chapter_6/Two_Photon_Imaging.html">
   6. Two Photon Imaging
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Chapter_7/Large_Scale_Electrophysiology.html">
   7. Large Scale Electrophysiology
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Chapter_8/Connectivity_Atlas.html">
   8. Allen Mouse Brain Connectivity Atlas
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Chapter_9/Brain_Imaging_Data_Structures.html">
   9. Brain Imaging Data Structures
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Chapter_10/Text_Analyses_of_Neuroscience_Literature.html">
   10. Text Analyses of Neuroscience Literature
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Chapter_11/Approaches_to_Analyzing_Large_Datasets.html">
   11. Approaches to Analyzing Large Datasets
  </a>
 </li>
</ul>

</nav>

 <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        <div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    
    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/Chapter_2/Pandas.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
    
</div>
        <!-- Source interaction buttons -->


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/NeuralDataScience/NeuralDataScience.github.io/master?urlpath=tree/Chapter_2/Pandas.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/NeuralDataScience/NeuralDataScience.github.io/blob/master/Chapter_2/Pandas.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-and-manipulate-dataframes">
   2.2.1. Create and Manipulate Dataframes
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#indexing-dataframes">
   2.2.2. Indexing Dataframes
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#subsetting">
   2.2.3. Subsetting
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#dataframe-methods">
   2.2.4. Dataframe Methods
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#additional-resources">
   2.2.5. Additional resources
  </a>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="pandas">
<h1><span class="section-number">2.2. </span>Pandas<a class="headerlink" href="#pandas" title="Permalink to this headline">¶</a></h1>
<p>In this chapter we will be discussing the advatages to using the <em>Pandas</em> package to analyze large datasets. While NumPy is a useful package, it can only be used with data of the same datatype. NumPy also does not support axis lables. <a class="reference external" href="https://pandas.pydata.org/docs/user_guide/index.html#user-guide">Pandas</a> is a package that helps us manipulate and analyze heterogenous data.</p>
<p>We strongly recommend looking at <a class="reference external" href="https://pandas.pydata.org/docs/user_guide/10min.html">“10 minutes to pandas”</a> for a broader overview, but here we’ll introduce the main concepts needed for the activities in this textbook.</p>
<p>Before we can use pandas, we need to import it. We can also nickname the modules when we import them. The convention is to import <code class="docutils literal notranslate"><span class="pre">pandas</span></code> as <code class="docutils literal notranslate"><span class="pre">pd</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import packages</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Use whos &#39;magic command&#39; to see available modules</span>
<span class="o">%</span><span class="k">whos</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Variable   Type      Data/Info
------------------------------
pd         module    &lt;module &#39;pandas&#39; from &#39;/U&lt;...&gt;ages/pandas/__init__.py&#39;&gt;
</pre></div>
</div>
</div>
</div>
<div class="section" id="create-and-manipulate-dataframes">
<h2><span class="section-number">2.2.1. </span>Create and Manipulate Dataframes<a class="headerlink" href="#create-and-manipulate-dataframes" title="Permalink to this headline">¶</a></h2>
<p>The two data structures of Pandas are the <code class="docutils literal notranslate"><span class="pre">series</span></code> and the <code class="docutils literal notranslate"><span class="pre">dataframe</span></code>. A <code class="docutils literal notranslate"><span class="pre">series</span></code> is a one-dimensional onject similar to a list. A <code class="docutils literal notranslate"><span class="pre">dataframe</span></code> can be thought of as a two-dimensional numpy array or a collection of <code class="docutils literal notranslate"><span class="pre">series</span></code> objects. Dataframes and Series can contain multiple different data types such as integers, strings, and floats, similar to an excel spreadsheet, unlike numpy arrays which can only contain values that are the same data type. Pandas also supports <code class="docutils literal notranslate"><span class="pre">string</span></code> lables unlike numpy arrays which only have numeric labels for their rows and columns. For a more in depth explanation, please visit the <a href = https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html> Intro to Data Structures</a> section in the Pandas User Guide.</p>
<p>You can create a Pandas dataframe by inputting dictionaries into the Pandas funtion <code class="docutils literal notranslate"><span class="pre">pd.DataFrame()</span></code>, by reading files, or through functions built into the Pandas package. The function <code class="docutils literal notranslate"><span class="pre">pd.read_csv()</span></code> reads a comma-seperated file and returns it as a <code class="docutils literal notranslate"><span class="pre">dataframe</span></code>. For more information on <code class="docutils literal notranslate"><span class="pre">pd.read_csv()</span></code> please visit the <a href = "https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html">documentation</a>.</p>
<p>Below we will create a dataframe by reading the file <code class="docutils literal notranslate"><span class="pre">brainarea_vs_genes_exp_w_reannotations-Copy1.tsv</span></code> which contains information on gene expression accross multiple brain areas. This dataset was created by Derek Howard and Abigail Mayes for the purpose of accelerating advances in data mining of open brain transcriptome data for polygenetic brain disorders. The data comes from normalized microarray datasets of gene expression from 6 adult human brains that was released by the Allen Brain Atlas and then processed into the dataframe we will see below. For more information on this dataset please visit the <a href = "https://github.com/derekhoward/HBAsets"> HBAsets repository</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in the list of lists as a data frame</span>
<span class="n">file_name</span> <span class="o">=</span> <span class="s1">&#39;brainarea_vs_genes_exp_w_reannotations-Copy1.tsv&#39;</span>
<span class="n">gene_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="c1"># &#39;.head()&#39; returns the first 5 rows in the dataframe</span>
<span class="n">gene_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
      <th>Crus I, lateral hemisphere</th>
      <th>Crus I, paravermis</th>
      <th>Crus II, lateral hemisphere</th>
      <th>Crus II, paravermis</th>
      <th>Edinger-Westphal nucleus</th>
      <th>...</th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
      <th>transverse gyri</th>
      <th>trochlear nucleus</th>
      <th>tuberomammillary nucleus</th>
      <th>ventral tegmental area</th>
      <th>ventromedial hypothalamic nucleus</th>
      <th>vestibular nuclei</th>
      <th>zona incerta</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>A1BG</td>
      <td>0.856487</td>
      <td>-1.773695</td>
      <td>-0.678679</td>
      <td>-0.986914</td>
      <td>0.826986</td>
      <td>0.948039</td>
      <td>0.935427</td>
      <td>1.120774</td>
      <td>-1.018554</td>
      <td>...</td>
      <td>0.277830</td>
      <td>0.514923</td>
      <td>0.733368</td>
      <td>-0.104286</td>
      <td>-0.910245</td>
      <td>1.039610</td>
      <td>-0.155167</td>
      <td>-0.444398</td>
      <td>-0.901361</td>
      <td>-0.236790</td>
    </tr>
    <tr>
      <th>1</th>
      <td>A1BG-AS1</td>
      <td>0.257664</td>
      <td>-1.373085</td>
      <td>-0.619923</td>
      <td>-0.636275</td>
      <td>0.362799</td>
      <td>0.353296</td>
      <td>0.422766</td>
      <td>0.346853</td>
      <td>-0.812015</td>
      <td>...</td>
      <td>1.074116</td>
      <td>0.821031</td>
      <td>1.219272</td>
      <td>0.901213</td>
      <td>-1.522431</td>
      <td>0.598719</td>
      <td>-1.709745</td>
      <td>-0.054156</td>
      <td>-1.695843</td>
      <td>-1.155961</td>
    </tr>
    <tr>
      <th>2</th>
      <td>A1CF</td>
      <td>-0.089614</td>
      <td>-0.546903</td>
      <td>0.282914</td>
      <td>-0.528926</td>
      <td>0.507916</td>
      <td>0.577696</td>
      <td>0.647671</td>
      <td>0.306824</td>
      <td>0.089958</td>
      <td>...</td>
      <td>-0.030265</td>
      <td>-0.187367</td>
      <td>-0.428358</td>
      <td>-0.465863</td>
      <td>-0.136936</td>
      <td>1.229487</td>
      <td>-0.110680</td>
      <td>-0.118175</td>
      <td>-0.139776</td>
      <td>0.123829</td>
    </tr>
    <tr>
      <th>3</th>
      <td>A2M</td>
      <td>0.552415</td>
      <td>-0.635485</td>
      <td>-0.954995</td>
      <td>-0.259745</td>
      <td>-1.687391</td>
      <td>-1.756847</td>
      <td>-1.640242</td>
      <td>-1.733110</td>
      <td>-0.091695</td>
      <td>...</td>
      <td>-0.058505</td>
      <td>0.207109</td>
      <td>-0.161808</td>
      <td>0.183630</td>
      <td>0.948098</td>
      <td>-0.977692</td>
      <td>0.911896</td>
      <td>-0.499357</td>
      <td>1.469386</td>
      <td>0.557998</td>
    </tr>
    <tr>
      <th>4</th>
      <td>A2ML1</td>
      <td>0.758031</td>
      <td>1.549857</td>
      <td>1.262225</td>
      <td>1.338780</td>
      <td>-0.289888</td>
      <td>-0.407026</td>
      <td>-0.358798</td>
      <td>-0.589988</td>
      <td>0.944684</td>
      <td>...</td>
      <td>-0.472908</td>
      <td>-0.598317</td>
      <td>-0.247797</td>
      <td>-0.282673</td>
      <td>1.396365</td>
      <td>0.945043</td>
      <td>0.158202</td>
      <td>0.572771</td>
      <td>0.073088</td>
      <td>-0.886780</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 233 columns</p>
</div></div></div>
</div>
<p>At the moment, our rows and columns don’t have any useful information – they’re simply a list of indices. We can reassign the row values by using the method <code class="docutils literal notranslate"><span class="pre">set_index()</span></code>. We can choose any column in our present dataframe to be the row values. Let’s assign the row lables to be the <code class="docutils literal notranslate"><span class="pre">gene_symbol</span></code> and reassign the dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">row_index</span> <span class="o">=</span> <span class="s1">&#39;gene_symbol&#39;</span>
<span class="n">gene_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">row_index</span><span class="p">)</span>
<span class="n">gene_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
      <th>Crus I, lateral hemisphere</th>
      <th>Crus I, paravermis</th>
      <th>Crus II, lateral hemisphere</th>
      <th>Crus II, paravermis</th>
      <th>Edinger-Westphal nucleus</th>
      <th>Heschl's gyrus</th>
      <th>...</th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
      <th>transverse gyri</th>
      <th>trochlear nucleus</th>
      <th>tuberomammillary nucleus</th>
      <th>ventral tegmental area</th>
      <th>ventromedial hypothalamic nucleus</th>
      <th>vestibular nuclei</th>
      <th>zona incerta</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>0.856487</td>
      <td>-1.773695</td>
      <td>-0.678679</td>
      <td>-0.986914</td>
      <td>0.826986</td>
      <td>0.948039</td>
      <td>0.935427</td>
      <td>1.120774</td>
      <td>-1.018554</td>
      <td>0.170282</td>
      <td>...</td>
      <td>0.277830</td>
      <td>0.514923</td>
      <td>0.733368</td>
      <td>-0.104286</td>
      <td>-0.910245</td>
      <td>1.039610</td>
      <td>-0.155167</td>
      <td>-0.444398</td>
      <td>-0.901361</td>
      <td>-0.236790</td>
    </tr>
    <tr>
      <th>A1BG-AS1</th>
      <td>0.257664</td>
      <td>-1.373085</td>
      <td>-0.619923</td>
      <td>-0.636275</td>
      <td>0.362799</td>
      <td>0.353296</td>
      <td>0.422766</td>
      <td>0.346853</td>
      <td>-0.812015</td>
      <td>0.903358</td>
      <td>...</td>
      <td>1.074116</td>
      <td>0.821031</td>
      <td>1.219272</td>
      <td>0.901213</td>
      <td>-1.522431</td>
      <td>0.598719</td>
      <td>-1.709745</td>
      <td>-0.054156</td>
      <td>-1.695843</td>
      <td>-1.155961</td>
    </tr>
    <tr>
      <th>A1CF</th>
      <td>-0.089614</td>
      <td>-0.546903</td>
      <td>0.282914</td>
      <td>-0.528926</td>
      <td>0.507916</td>
      <td>0.577696</td>
      <td>0.647671</td>
      <td>0.306824</td>
      <td>0.089958</td>
      <td>0.149820</td>
      <td>...</td>
      <td>-0.030265</td>
      <td>-0.187367</td>
      <td>-0.428358</td>
      <td>-0.465863</td>
      <td>-0.136936</td>
      <td>1.229487</td>
      <td>-0.110680</td>
      <td>-0.118175</td>
      <td>-0.139776</td>
      <td>0.123829</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>0.552415</td>
      <td>-0.635485</td>
      <td>-0.954995</td>
      <td>-0.259745</td>
      <td>-1.687391</td>
      <td>-1.756847</td>
      <td>-1.640242</td>
      <td>-1.733110</td>
      <td>-0.091695</td>
      <td>0.003428</td>
      <td>...</td>
      <td>-0.058505</td>
      <td>0.207109</td>
      <td>-0.161808</td>
      <td>0.183630</td>
      <td>0.948098</td>
      <td>-0.977692</td>
      <td>0.911896</td>
      <td>-0.499357</td>
      <td>1.469386</td>
      <td>0.557998</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>0.758031</td>
      <td>1.549857</td>
      <td>1.262225</td>
      <td>1.338780</td>
      <td>-0.289888</td>
      <td>-0.407026</td>
      <td>-0.358798</td>
      <td>-0.589988</td>
      <td>0.944684</td>
      <td>-0.466327</td>
      <td>...</td>
      <td>-0.472908</td>
      <td>-0.598317</td>
      <td>-0.247797</td>
      <td>-0.282673</td>
      <td>1.396365</td>
      <td>0.945043</td>
      <td>0.158202</td>
      <td>0.572771</td>
      <td>0.073088</td>
      <td>-0.886780</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 232 columns</p>
</div></div></div>
</div>
<p>It would help to know what information is in our dataset. In other words, what is across the columns at the top? We can get a list by accessing the <code class="docutils literal notranslate"><span class="pre">columns</span></code> attribute.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Access the columns of our dataframe </span>
<span class="n">gene_df_columns</span> <span class="o">=</span> <span class="n">gene_df</span><span class="o">.</span><span class="n">columns</span> 
<span class="n">gene_df_columns</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;CA1 field&#39;, &#39;CA2 field&#39;, &#39;CA3 field&#39;, &#39;CA4 field&#39;,
       &#39;Crus I, lateral hemisphere&#39;, &#39;Crus I, paravermis&#39;,
       &#39;Crus II, lateral hemisphere&#39;, &#39;Crus II, paravermis&#39;,
       &#39;Edinger-Westphal nucleus&#39;, &#39;Heschl&#39;s gyrus&#39;,
       ...
       &#39;temporal pole, inferior aspect&#39;, &#39;temporal pole, medial aspect&#39;,
       &#39;temporal pole, superior aspect&#39;, &#39;transverse gyri&#39;,
       &#39;trochlear nucleus&#39;, &#39;tuberomammillary nucleus&#39;,
       &#39;ventral tegmental area&#39;, &#39;ventromedial hypothalamic nucleus&#39;,
       &#39;vestibular nuclei&#39;, &#39;zona incerta&#39;],
      dtype=&#39;object&#39;, length=232)
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="indexing-dataframes">
<h2><span class="section-number">2.2.2. </span>Indexing Dataframes<a class="headerlink" href="#indexing-dataframes" title="Permalink to this headline">¶</a></h2>
<p>Indexing in Pandas works slightly different than in NumPy. Similar to a dictionary, we can index Dataframes by their names instead of using the specific index number like in NumPy.</p>
<p>The syntax for indexing Dataframes is <code class="docutils literal notranslate"><span class="pre">dataframe.loc[row_label,column_label]</span></code>. To index an individual column, we use the shorthand syntax <code class="docutils literal notranslate"><span class="pre">df[column_label]</span></code>. To index an individual row, we use the syntax <code class="docutils literal notranslate"><span class="pre">df.loc[row_label]</span></code>. To index by index #, we use the syntax <code class="docutils literal notranslate"><span class="pre">df.iloc[index_number]</span></code>. Below are some examples on how to access rows, columns, and single values in our dataframe. For more information on indexing dataframes, visit the <a href = "https://pandas.pydata.org/docs/user_guide/indexing.html#indexing"> “Indexing and selecting data”</a> section in the Pandas User Guide.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select a single column</span>
<span class="n">column</span> <span class="o">=</span> <span class="s1">&#39;CA1 field&#39;</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gene expression values in CA1 field:&#39;</span><span class="p">)</span>
<span class="n">gene_df</span><span class="p">[</span><span class="n">column</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene expression values in CA1 field:
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>gene_symbol
A1BG        0.856487
A1BG-AS1    0.257664
A1CF       -0.089614
A2M         0.552415
A2ML1       0.758031
              ...   
ZYG11A     -0.496398
ZYG11B     -0.856866
ZYX        -1.941816
ZZEF1      -0.015748
ZZZ3       -0.924901
Name: CA1 field, Length: 20869, dtype: float64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select a single row</span>
<span class="n">row</span> <span class="o">=</span> <span class="s1">&#39;A1BG&#39;</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gene expression of A1BG across brain regeions:&#39;</span><span class="p">)</span>
<span class="n">gene_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">row</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene expression of A1BG across brain regeions:
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CA1 field                            0.856487
CA2 field                           -1.773695
CA3 field                           -0.678679
CA4 field                           -0.986914
Crus I, lateral hemisphere           0.826986
                                       ...   
tuberomammillary nucleus             1.039610
ventral tegmental area              -0.155167
ventromedial hypothalamic nucleus   -0.444398
vestibular nuclei                   -0.901361
zona incerta                        -0.236790
Name: A1BG, Length: 232, dtype: float64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select an individual value </span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gene expression of A1BG in CA1 field:&#39;</span><span class="p">)</span>
<span class="n">gene_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">row</span><span class="p">,</span> <span class="n">column</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene expression of A1BG in CA1 field:
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.8564873784944677
</pre></div>
</div>
</div>
</div>
<p>To select multiple different columns, you would create a <code class="docutils literal notranslate"><span class="pre">list</span></code> of all your columns of interest as so:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select multiple columns</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gene expression values in CA1, CA3, and Crus I, lateral hemisphere :&#39;</span><span class="p">)</span>
<span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CA1 field&#39;</span><span class="p">,</span> <span class="s1">&#39;CA3 field&#39;</span><span class="p">,</span> <span class="s1">&#39;Crus I, lateral hemisphere&#39;</span><span class="p">]</span>
<span class="n">gene_df</span><span class="p">[</span><span class="n">columns</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene expression values in CA1, CA3, and Crus I, lateral hemisphere :
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CA1 field</th>
      <th>CA3 field</th>
      <th>Crus I, lateral hemisphere</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>0.856487</td>
      <td>-0.678679</td>
      <td>0.826986</td>
    </tr>
    <tr>
      <th>A1BG-AS1</th>
      <td>0.257664</td>
      <td>-0.619923</td>
      <td>0.362799</td>
    </tr>
    <tr>
      <th>A1CF</th>
      <td>-0.089614</td>
      <td>0.282914</td>
      <td>0.507916</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>0.552415</td>
      <td>-0.954995</td>
      <td>-1.687391</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>0.758031</td>
      <td>1.262225</td>
      <td>-0.289888</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>ZYG11A</th>
      <td>-0.496398</td>
      <td>0.325555</td>
      <td>0.158885</td>
    </tr>
    <tr>
      <th>ZYG11B</th>
      <td>-0.856866</td>
      <td>0.701878</td>
      <td>0.337138</td>
    </tr>
    <tr>
      <th>ZYX</th>
      <td>-1.941816</td>
      <td>-0.681255</td>
      <td>0.872683</td>
    </tr>
    <tr>
      <th>ZZEF1</th>
      <td>-0.015748</td>
      <td>0.743609</td>
      <td>1.108376</td>
    </tr>
    <tr>
      <th>ZZZ3</th>
      <td>-0.924901</td>
      <td>0.108320</td>
      <td>-1.591413</td>
    </tr>
  </tbody>
</table>
<p>20869 rows × 3 columns</p>
</div></div></div>
</div>
</div>
<div class="section" id="subsetting">
<h2><span class="section-number">2.2.3. </span>Subsetting<a class="headerlink" href="#subsetting" title="Permalink to this headline">¶</a></h2>
<p>Like numpy arrays, subet our original dataframe to only include data that meets our criteria. Our dataframe has data on multiple different brain areas with many gene expression values. Filter your dataframe by using the following syntax:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">new_df</span> <span class="o">=</span> <span class="n">original_df</span><span class="p">[</span><span class="n">original_df</span><span class="p">[</span><span class="s1">&#39;Column of Interest&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Desired Value&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>In plain english, what this is saying is: save a dataframe from the original dataframe, where the original dataframe values in my Column of Interest are equal to my Desired Value. For more inofrmation on subsetting,  visit the <a href = "https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html"> “How do I selcect a subset a subset of a Dataframe”</a> section in the Pandas documentation.</p>
<p>Below we will demonstrate how to execute this by taking a look at the <code class="docutils literal notranslate"><span class="pre">CA1</span> <span class="pre">field</span></code> column in <code class="docutils literal notranslate"><span class="pre">gene_df</span></code>. We will create a dataframe from <code class="docutils literal notranslate"><span class="pre">gene_df</span></code> that only contains genes that showed a certain level of gene expression.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a dataframe with only genes that have an expression </span>
<span class="c1"># value greater than 1.7 in &#39;CA1 field&#39; </span>
<span class="n">desired_column</span> <span class="o">=</span> <span class="s1">&#39;CA1 field&#39;</span>
<span class="n">desired_value</span> <span class="o">=</span> <span class="mf">1.7</span>
<span class="n">new_gene_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[</span><span class="n">gene_df</span><span class="p">[</span><span class="n">desired_column</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">desired_value</span><span class="p">]</span>
<span class="n">new_gene_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
      <th>Crus I, lateral hemisphere</th>
      <th>Crus I, paravermis</th>
      <th>Crus II, lateral hemisphere</th>
      <th>Crus II, paravermis</th>
      <th>Edinger-Westphal nucleus</th>
      <th>Heschl's gyrus</th>
      <th>...</th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
      <th>transverse gyri</th>
      <th>trochlear nucleus</th>
      <th>tuberomammillary nucleus</th>
      <th>ventral tegmental area</th>
      <th>ventromedial hypothalamic nucleus</th>
      <th>vestibular nuclei</th>
      <th>zona incerta</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ABCC12</th>
      <td>2.089999</td>
      <td>0.684837</td>
      <td>0.097313</td>
      <td>-0.051411</td>
      <td>-1.078900</td>
      <td>-0.912071</td>
      <td>-1.131497</td>
      <td>-0.799075</td>
      <td>-0.009423</td>
      <td>0.889132</td>
      <td>...</td>
      <td>1.289340</td>
      <td>0.885886</td>
      <td>1.271053</td>
      <td>0.650801</td>
      <td>-0.083413</td>
      <td>-0.793237</td>
      <td>-0.499512</td>
      <td>-0.762330</td>
      <td>-0.902496</td>
      <td>-0.904421</td>
    </tr>
    <tr>
      <th>ABHD17C</th>
      <td>1.716973</td>
      <td>0.601041</td>
      <td>1.132500</td>
      <td>1.354679</td>
      <td>-0.923195</td>
      <td>-0.887576</td>
      <td>-1.122027</td>
      <td>-0.926876</td>
      <td>0.339678</td>
      <td>0.208174</td>
      <td>...</td>
      <td>0.530983</td>
      <td>0.809101</td>
      <td>0.763170</td>
      <td>-0.093598</td>
      <td>-0.313468</td>
      <td>-0.164013</td>
      <td>-0.537971</td>
      <td>1.169105</td>
      <td>-0.663245</td>
      <td>-0.377181</td>
    </tr>
    <tr>
      <th>ABI1</th>
      <td>2.051762</td>
      <td>2.571777</td>
      <td>2.472188</td>
      <td>2.261170</td>
      <td>0.366138</td>
      <td>0.507783</td>
      <td>0.449606</td>
      <td>0.498424</td>
      <td>-1.365242</td>
      <td>0.472104</td>
      <td>...</td>
      <td>0.868471</td>
      <td>1.197022</td>
      <td>0.914783</td>
      <td>0.473846</td>
      <td>-2.230209</td>
      <td>-0.330684</td>
      <td>-1.153189</td>
      <td>-0.073650</td>
      <td>-1.681602</td>
      <td>-1.049258</td>
    </tr>
    <tr>
      <th>ACTB</th>
      <td>2.489711</td>
      <td>2.806688</td>
      <td>2.461655</td>
      <td>2.340131</td>
      <td>-1.296731</td>
      <td>-1.334696</td>
      <td>-1.158460</td>
      <td>-1.461027</td>
      <td>0.088209</td>
      <td>0.027236</td>
      <td>...</td>
      <td>-0.370215</td>
      <td>-0.946920</td>
      <td>-0.197363</td>
      <td>-0.094468</td>
      <td>0.926183</td>
      <td>-0.034337</td>
      <td>0.803164</td>
      <td>-0.389238</td>
      <td>0.596916</td>
      <td>0.110921</td>
    </tr>
    <tr>
      <th>ACTR2</th>
      <td>2.655049</td>
      <td>2.384445</td>
      <td>1.728348</td>
      <td>1.413585</td>
      <td>-1.377474</td>
      <td>-1.207922</td>
      <td>-1.496018</td>
      <td>-1.359462</td>
      <td>0.632764</td>
      <td>0.259402</td>
      <td>...</td>
      <td>0.215683</td>
      <td>0.142880</td>
      <td>0.080649</td>
      <td>0.312218</td>
      <td>1.793563</td>
      <td>-0.028263</td>
      <td>-0.189930</td>
      <td>-0.389064</td>
      <td>-0.484580</td>
      <td>-1.478232</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 232 columns</p>
</div></div></div>
</div>
</div>
<div class="section" id="dataframe-methods">
<h2><span class="section-number">2.2.4. </span>Dataframe Methods<a class="headerlink" href="#dataframe-methods" title="Permalink to this headline">¶</a></h2>
<p>Pandas has many useful methods that you can use on your data, including <code class="docutils literal notranslate"><span class="pre">describe</span></code>, <code class="docutils literal notranslate"><span class="pre">mean</span></code>, and more. To learn more about all the different methods that can be used to manipulate and analyze dataframes, please visit the <a href = "https://pandas.pydata.org/docs/user_guide/index.html"> Pandas User Guide </a>. We will demonstrate some of these methods below.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">describe</span></code> method returns descriptive statistics of all the columns in our dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_df</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
      <th>Crus I, lateral hemisphere</th>
      <th>Crus I, paravermis</th>
      <th>Crus II, lateral hemisphere</th>
      <th>Crus II, paravermis</th>
      <th>Edinger-Westphal nucleus</th>
      <th>Heschl's gyrus</th>
      <th>...</th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
      <th>transverse gyri</th>
      <th>trochlear nucleus</th>
      <th>tuberomammillary nucleus</th>
      <th>ventral tegmental area</th>
      <th>ventromedial hypothalamic nucleus</th>
      <th>vestibular nuclei</th>
      <th>zona incerta</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>...</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>0.003664</td>
      <td>0.017002</td>
      <td>0.015315</td>
      <td>-0.016633</td>
      <td>0.093686</td>
      <td>0.088810</td>
      <td>0.096118</td>
      <td>0.087608</td>
      <td>0.047124</td>
      <td>-0.042517</td>
      <td>...</td>
      <td>-0.056050</td>
      <td>-0.051731</td>
      <td>-0.049905</td>
      <td>-0.039528</td>
      <td>0.059726</td>
      <td>0.014856</td>
      <td>0.009535</td>
      <td>-0.002853</td>
      <td>-0.002527</td>
      <td>0.013018</td>
    </tr>
    <tr>
      <th>std</th>
      <td>0.924456</td>
      <td>1.129368</td>
      <td>1.078987</td>
      <td>0.897192</td>
      <td>1.146146</td>
      <td>1.118501</td>
      <td>1.172986</td>
      <td>1.136823</td>
      <td>0.973224</td>
      <td>0.500526</td>
      <td>...</td>
      <td>0.567950</td>
      <td>0.651098</td>
      <td>0.636729</td>
      <td>0.494012</td>
      <td>1.158916</td>
      <td>0.897387</td>
      <td>0.686320</td>
      <td>0.830602</td>
      <td>0.723977</td>
      <td>0.725770</td>
    </tr>
    <tr>
      <th>min</th>
      <td>-4.076424</td>
      <td>-5.923691</td>
      <td>-5.994731</td>
      <td>-3.971984</td>
      <td>-2.739924</td>
      <td>-2.662897</td>
      <td>-2.908676</td>
      <td>-2.864308</td>
      <td>-3.671242</td>
      <td>-1.666268</td>
      <td>...</td>
      <td>-1.840486</td>
      <td>-2.433961</td>
      <td>-2.412614</td>
      <td>-1.655962</td>
      <td>-6.330275</td>
      <td>-3.141490</td>
      <td>-1.977225</td>
      <td>-3.541112</td>
      <td>-2.369304</td>
      <td>-2.348784</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>-0.570475</td>
      <td>-0.644093</td>
      <td>-0.631248</td>
      <td>-0.573605</td>
      <td>-0.802651</td>
      <td>-0.778933</td>
      <td>-0.824404</td>
      <td>-0.794683</td>
      <td>-0.651691</td>
      <td>-0.414882</td>
      <td>...</td>
      <td>-0.472481</td>
      <td>-0.517562</td>
      <td>-0.529458</td>
      <td>-0.404416</td>
      <td>-0.751329</td>
      <td>-0.602619</td>
      <td>-0.497988</td>
      <td>-0.542164</td>
      <td>-0.532177</td>
      <td>-0.493921</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>-0.025821</td>
      <td>0.011189</td>
      <td>0.006140</td>
      <td>-0.044566</td>
      <td>0.100558</td>
      <td>0.098665</td>
      <td>0.109358</td>
      <td>0.096585</td>
      <td>0.022760</td>
      <td>-0.057495</td>
      <td>...</td>
      <td>-0.091029</td>
      <td>-0.083971</td>
      <td>-0.089572</td>
      <td>-0.051758</td>
      <td>0.037410</td>
      <td>-0.030474</td>
      <td>-0.023799</td>
      <td>-0.023471</td>
      <td>-0.045003</td>
      <td>-0.031623</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>0.561571</td>
      <td>0.706946</td>
      <td>0.706398</td>
      <td>0.547903</td>
      <td>0.985159</td>
      <td>0.951919</td>
      <td>1.007791</td>
      <td>0.960813</td>
      <td>0.736614</td>
      <td>0.312739</td>
      <td>...</td>
      <td>0.352796</td>
      <td>0.412321</td>
      <td>0.403382</td>
      <td>0.310551</td>
      <td>0.844723</td>
      <td>0.587382</td>
      <td>0.503613</td>
      <td>0.508410</td>
      <td>0.501191</td>
      <td>0.492262</td>
    </tr>
    <tr>
      <th>max</th>
      <td>7.062717</td>
      <td>7.387742</td>
      <td>6.413603</td>
      <td>7.178692</td>
      <td>2.679149</td>
      <td>2.717237</td>
      <td>2.963899</td>
      <td>2.857205</td>
      <td>7.552440</td>
      <td>2.234669</td>
      <td>...</td>
      <td>2.199291</td>
      <td>2.631498</td>
      <td>3.065735</td>
      <td>2.238555</td>
      <td>6.892682</td>
      <td>5.968364</td>
      <td>7.267837</td>
      <td>6.650673</td>
      <td>2.723777</td>
      <td>2.845665</td>
    </tr>
  </tbody>
</table>
<p>8 rows × 232 columns</p>
</div></div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">mean</span></code> and <code class="docutils literal notranslate"><span class="pre">std</span></code> method return the mean and standard deviation of each column in the dataframe, respectfully.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_df</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CA1 field                            0.003664
CA2 field                            0.017002
CA3 field                            0.015315
CA4 field                           -0.016633
Crus I, lateral hemisphere           0.093686
                                       ...   
tuberomammillary nucleus             0.014856
ventral tegmental area               0.009535
ventromedial hypothalamic nucleus   -0.002853
vestibular nuclei                   -0.002527
zona incerta                         0.013018
Length: 232, dtype: float64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_df</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CA1 field                            0.924456
CA2 field                            1.129368
CA3 field                            1.078987
CA4 field                            0.897192
Crus I, lateral hemisphere           1.146146
                                       ...   
tuberomammillary nucleus             0.897387
ventral tegmental area               0.686320
ventromedial hypothalamic nucleus    0.830602
vestibular nuclei                    0.723977
zona incerta                         0.725770
Length: 232, dtype: float64
</pre></div>
</div>
</div>
</div>
<p>Let’s say we have two different dataframes and we would like to combine the two into one single dataframe. We can use either the <code class="docutils literal notranslate"><span class="pre">merge</span></code> or <code class="docutils literal notranslate"><span class="pre">join</span></code> Pandas methods in order to pull all of this data into one dataframe.</p>
<p><img alt="" src="http://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png" /></p>
<p>There are different types of joins/merges you can do in Pandas, illustrated <a href="http://www.datasciencemadesimple.com/join-merge-data-frames-pandas-python/">above</a>. Here, we want to do an <strong>inner</strong> merge, where we’re only keeping entries with indices that are in both dataframes. We could do this merge based on columns, alternatively.</p>
<p><strong>Inner</strong> is the default kind of join, so we do not need to specify it. And by default, join will use the ‘left’ dataframe, in other words, the dataframe that is executing the <code class="docutils literal notranslate"><span class="pre">join</span></code> method.</p>
<p>If you need more information, look at the <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html">join</a> and <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html">merge</a> documentation: you can use either of these to unite your dataframes, though join will be simpler!</p>
<p>Below is an example of how to join two separate dataframe into one, unified dataframe. We start with one dataframe with only entries from the <em>temporal pole</em> and another dataframe with only entries from the CA fields of the hippocampus. We can then join the two dataframes together using the syntax <code class="docutils literal notranslate"><span class="pre">unified_df</span> <span class="pre">=</span> <span class="pre">df_1.join(df_2)</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dataframe w/ only Temporal Pole entries </span>
<span class="n">temporal_pole_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[[</span><span class="s1">&#39;temporal pole, inferior aspect&#39;</span><span class="p">,</span> 
                            <span class="s1">&#39;temporal pole, medial aspect&#39;</span><span class="p">,</span> 
                            <span class="s1">&#39;temporal pole, superior aspect&#39;</span><span class="p">]]</span>
<span class="n">temporal_pole_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>0.277830</td>
      <td>0.514923</td>
      <td>0.733368</td>
    </tr>
    <tr>
      <th>A1BG-AS1</th>
      <td>1.074116</td>
      <td>0.821031</td>
      <td>1.219272</td>
    </tr>
    <tr>
      <th>A1CF</th>
      <td>-0.030265</td>
      <td>-0.187367</td>
      <td>-0.428358</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>-0.058505</td>
      <td>0.207109</td>
      <td>-0.161808</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>-0.472908</td>
      <td>-0.598317</td>
      <td>-0.247797</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dataframe w/ only CA field entries </span>
<span class="n">CA_field_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[[</span><span class="s1">&#39;CA1 field&#39;</span><span class="p">,</span> 
                       <span class="s1">&#39;CA2 field&#39;</span><span class="p">,</span> 
                       <span class="s1">&#39;CA3 field&#39;</span><span class="p">,</span> 
                       <span class="s1">&#39;CA4 field&#39;</span><span class="p">]]</span>
<span class="n">CA_field_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>0.856487</td>
      <td>-1.773695</td>
      <td>-0.678679</td>
      <td>-0.986914</td>
    </tr>
    <tr>
      <th>A1BG-AS1</th>
      <td>0.257664</td>
      <td>-1.373085</td>
      <td>-0.619923</td>
      <td>-0.636275</td>
    </tr>
    <tr>
      <th>A1CF</th>
      <td>-0.089614</td>
      <td>-0.546903</td>
      <td>0.282914</td>
      <td>-0.528926</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>0.552415</td>
      <td>-0.635485</td>
      <td>-0.954995</td>
      <td>-0.259745</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>0.758031</td>
      <td>1.549857</td>
      <td>1.262225</td>
      <td>1.338780</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Merge/Join the two dataframes</span>
<span class="n">df_1</span> <span class="o">=</span> <span class="n">temporal_pole_df</span>
<span class="n">df_2</span> <span class="o">=</span> <span class="n">CA_field_df</span>

<span class="n">unified_df</span> <span class="o">=</span> <span class="n">df_1</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">df_2</span><span class="p">)</span>
<span class="n">unified_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>0.277830</td>
      <td>0.514923</td>
      <td>0.733368</td>
      <td>0.856487</td>
      <td>-1.773695</td>
      <td>-0.678679</td>
      <td>-0.986914</td>
    </tr>
    <tr>
      <th>A1BG-AS1</th>
      <td>1.074116</td>
      <td>0.821031</td>
      <td>1.219272</td>
      <td>0.257664</td>
      <td>-1.373085</td>
      <td>-0.619923</td>
      <td>-0.636275</td>
    </tr>
    <tr>
      <th>A1CF</th>
      <td>-0.030265</td>
      <td>-0.187367</td>
      <td>-0.428358</td>
      <td>-0.089614</td>
      <td>-0.546903</td>
      <td>0.282914</td>
      <td>-0.528926</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>-0.058505</td>
      <td>0.207109</td>
      <td>-0.161808</td>
      <td>0.552415</td>
      <td>-0.635485</td>
      <td>-0.954995</td>
      <td>-0.259745</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>-0.472908</td>
      <td>-0.598317</td>
      <td>-0.247797</td>
      <td>0.758031</td>
      <td>1.549857</td>
      <td>1.262225</td>
      <td>1.338780</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</div>
<div class="section" id="additional-resources">
<h2><span class="section-number">2.2.5. </span>Additional resources<a class="headerlink" href="#additional-resources" title="Permalink to this headline">¶</a></h2>
<p>See the <a class="reference external" href="https://jakevdp.github.io/PythonDataScienceHandbook/03.00-introduction-to-pandas.html">Python Data Science Handbook</a> for a more in depth exploration of Pandas, and of course, the original documentation.</p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./Chapter_2"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="NumPy.html" title="previous page"><span class="section-number">2.1. </span>NumPy</a>
    <a class='right-next' id="next-link" href="SciPy.html" title="next page"><span class="section-number">2.3. </span>SciPy</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Victor Magdaleno-Garcia<br/>
        
            &copy; Copyright 2020.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../_static/js/index.3da636dd464baa7582d2.js"></script>


    
  </body>
</html>