
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Pandas &#8212; Neuroscience in the Age of Data Science</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Chapter03/Pandas';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="SciPy" href="SciPy.html" />
    <link rel="prev" title="NumPy" href="NumPy.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/brainz.jpeg" class="logo__image only-light" alt="Neuroscience in the Age of Data Science - Home"/>
    <script>document.write(`<img src="../_static/brainz.jpeg" class="logo__image only-dark" alt="Neuroscience in the Age of Data Science - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../index.html">
                    Neuroscience in the Age of Data Science
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">DATA SCIENCE IN PYTHON</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="Data_Science_In_Python.html">Data Science in Python</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="NumPy.html">NumPy</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Pandas</a></li>
<li class="toctree-l2"><a class="reference internal" href="SciPy.html">SciPy</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">FINDING &amp; EXPLORING NEURAL DATA</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../Chapter04/FindingData.html">Finding Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Chapter05/DataExploration_BrainSize.html">Exploring the Data</a></li>

</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">KINDS OF NEURAL DATA</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../Chapter09/SingleCellEphys.html">Defining cell types by their electrophysiology</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Chapter09/2pImaging.html">Brain Observatory: <em>What are those cells doing?</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">APPROACHES TO WORKING WITH BIG DATA</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../Chapter16/DimensionalityReduction.html">Dimensionality Reduction &amp; Clustering</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/NeuralDataScience/NeuralDataScience.github.io/blob/master/Chapter03/Pandas.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch on Colab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img alt="Colab logo" src="../_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/Chapter03/Pandas.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Pandas</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#create-and-manipulate-dataframes">Create and Manipulate Dataframes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#dataframe-example">DataFrame example</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#indexing-dataframes">Indexing Dataframes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#subsetting">Subsetting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dataframe-methods">DataFrame Methods</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#additional-resources">Additional resources</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="pandas">
<h1>Pandas<a class="headerlink" href="#pandas" title="Link to this heading">#</a></h1>
<p>In this chapter we will be discussing the advantages to using the <em>Pandas</em> package to analyze large datasets. While NumPy is a useful package, it can only be used with data of the same datatype. <a class="reference external" href="https://pandas.pydata.org/docs/user_guide/index.html#user-guide">Pandas</a>, however, is a package that helps us manipulate and analyze heterogeneous data.</p>
<p>We strongly recommend looking at <a class="reference external" href="https://pandas.pydata.org/docs/user_guide/10min.html">“10 minutes to pandas”</a> for a broader overview, but here we’ll introduce the main concepts needed for the activities in this textbook.</p>
<p>Before we can use pandas, we need to import it. We can also nickname the modules when we import them. The convention is to import <code class="docutils literal notranslate"><span class="pre">pandas</span></code> as <code class="docutils literal notranslate"><span class="pre">pd</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import packages</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="c1"># Use whos &#39;magic command&#39; to see available modules</span>
<span class="o">%</span><span class="k">whos</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Variable   Type      Data/Info
------------------------------
pd         module    &lt;module &#39;pandas&#39; from &#39;/U&lt;...&gt;ages/pandas/__init__.py&#39;&gt;
</pre></div>
</div>
</div>
</div>
<section id="create-and-manipulate-dataframes">
<h2>Create and Manipulate Dataframes<a class="headerlink" href="#create-and-manipulate-dataframes" title="Link to this heading">#</a></h2>
<p>The two data structures of Pandas are the <code class="docutils literal notranslate"><span class="pre">Series</span></code> and the <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code>. A <code class="docutils literal notranslate"><span class="pre">Series</span></code> is a one-dimensional object similar to a list. A <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> can be thought of as a two-dimensional numpy array or a collection of <code class="docutils literal notranslate"><span class="pre">Series</span></code> objects. Series and dataframes can contain multiple different data types such as integers, strings, and floats, similar to an Excel spreadsheet. Pandas also supports <code class="docutils literal notranslate"><span class="pre">string</span></code> lables unlike numpy arrays which only have numeric labels for their rows and columns. For a more in depth explanation, please visit the <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html">Introduction to Data Structures</a> section in the Pandas User Guide.</p>
<p>You can create a Pandas dataframe by inputting dictionaries into the Pandas function <code class="docutils literal notranslate"><span class="pre">pd.DataFrame()</span></code>, by reading files, or through functions built into the Pandas package. The function <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html"><code class="docutils literal notranslate"><span class="pre">pd.read_csv()</span></code></a> reads a comma- or tab-separated file and returns it as a <code class="docutils literal notranslate"><span class="pre">dataframe</span></code>.</p>
<section id="dataframe-example">
<h3>DataFrame example<a class="headerlink" href="#dataframe-example" title="Link to this heading">#</a></h3>
<p>Below we will create a dataframe by reading the file <code class="docutils literal notranslate"><span class="pre">brainarea_vs_genes_exp_w_reannotations.tsv</span></code> which contains information on gene expression accross multiple brain areas.</p>
<blockquote>
<div><p><strong>About this dataset:</strong>
This dataset was created by Derek Howard and Abigail Mayes for the purpose of accelerating advances in data mining of open brain transcriptome data for polygenetic brain disorders. The data comes from normalized microarray datasets of gene expression from 6 adult human brains that was released by the Allen Brain Institute and then processed into the dataframe we will see below. For more information on this dataset please visit the <a href = "https://github.com/derekhoward/HBAsets"> HBAsets repository</a>.</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in the list of lists as a data frame</span>
<span class="n">file_name</span> <span class="o">=</span> <span class="s1">&#39;brainarea_vs_genes_exp_w_reannotations.tsv&#39;</span>
<span class="n">gene_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="c1"># &#39;.head()&#39; returns the first 5 rows in the dataframe</span>
<span class="n">gene_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
      <th>Crus I, lateral hemisphere</th>
      <th>Crus I, paravermis</th>
      <th>Crus II, lateral hemisphere</th>
      <th>Crus II, paravermis</th>
      <th>Edinger-Westphal nucleus</th>
      <th>...</th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
      <th>transverse gyri</th>
      <th>trochlear nucleus</th>
      <th>tuberomammillary nucleus</th>
      <th>ventral tegmental area</th>
      <th>ventromedial hypothalamic nucleus</th>
      <th>vestibular nuclei</th>
      <th>zona incerta</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>A1BG</td>
      <td>0.856487</td>
      <td>-1.773695</td>
      <td>-0.678679</td>
      <td>-0.986914</td>
      <td>0.826986</td>
      <td>0.948039</td>
      <td>0.935427</td>
      <td>1.120774</td>
      <td>-1.018554</td>
      <td>...</td>
      <td>0.277830</td>
      <td>0.514923</td>
      <td>0.733368</td>
      <td>-0.104286</td>
      <td>-0.910245</td>
      <td>1.039610</td>
      <td>-0.155167</td>
      <td>-0.444398</td>
      <td>-0.901361</td>
      <td>-0.236790</td>
    </tr>
    <tr>
      <th>1</th>
      <td>A1BG-AS1</td>
      <td>0.257664</td>
      <td>-1.373085</td>
      <td>-0.619923</td>
      <td>-0.636275</td>
      <td>0.362799</td>
      <td>0.353296</td>
      <td>0.422766</td>
      <td>0.346853</td>
      <td>-0.812015</td>
      <td>...</td>
      <td>1.074116</td>
      <td>0.821031</td>
      <td>1.219272</td>
      <td>0.901213</td>
      <td>-1.522431</td>
      <td>0.598719</td>
      <td>-1.709745</td>
      <td>-0.054156</td>
      <td>-1.695843</td>
      <td>-1.155961</td>
    </tr>
    <tr>
      <th>2</th>
      <td>A1CF</td>
      <td>-0.089614</td>
      <td>-0.546903</td>
      <td>0.282914</td>
      <td>-0.528926</td>
      <td>0.507916</td>
      <td>0.577696</td>
      <td>0.647671</td>
      <td>0.306824</td>
      <td>0.089958</td>
      <td>...</td>
      <td>-0.030265</td>
      <td>-0.187367</td>
      <td>-0.428358</td>
      <td>-0.465863</td>
      <td>-0.136936</td>
      <td>1.229487</td>
      <td>-0.110680</td>
      <td>-0.118175</td>
      <td>-0.139776</td>
      <td>0.123829</td>
    </tr>
    <tr>
      <th>3</th>
      <td>A2M</td>
      <td>0.552415</td>
      <td>-0.635485</td>
      <td>-0.954995</td>
      <td>-0.259745</td>
      <td>-1.687391</td>
      <td>-1.756847</td>
      <td>-1.640242</td>
      <td>-1.733110</td>
      <td>-0.091695</td>
      <td>...</td>
      <td>-0.058505</td>
      <td>0.207109</td>
      <td>-0.161808</td>
      <td>0.183630</td>
      <td>0.948098</td>
      <td>-0.977692</td>
      <td>0.911896</td>
      <td>-0.499357</td>
      <td>1.469386</td>
      <td>0.557998</td>
    </tr>
    <tr>
      <th>4</th>
      <td>A2ML1</td>
      <td>0.758031</td>
      <td>1.549857</td>
      <td>1.262225</td>
      <td>1.338780</td>
      <td>-0.289888</td>
      <td>-0.407026</td>
      <td>-0.358798</td>
      <td>-0.589988</td>
      <td>0.944684</td>
      <td>...</td>
      <td>-0.472908</td>
      <td>-0.598317</td>
      <td>-0.247797</td>
      <td>-0.282673</td>
      <td>1.396365</td>
      <td>0.945043</td>
      <td>0.158202</td>
      <td>0.572771</td>
      <td>0.073088</td>
      <td>-0.886780</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 233 columns</p>
</div></div></div>
</div>
<p>At the moment, the first column  of information above, the <strong>index</strong> just contains a list of numbers. We can reassign the row labels by using the method <code class="docutils literal notranslate"><span class="pre">set_index()</span></code>. We can choose any column in our present dataframe to be the row values. Let’s assign the row lables to be the <code class="docutils literal notranslate"><span class="pre">gene_symbol</span></code> and reassign the dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">row_index</span> <span class="o">=</span> <span class="s1">&#39;gene_symbol&#39;</span>
<span class="n">gene_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">row_index</span><span class="p">)</span>
<span class="n">gene_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
      <th>Crus I, lateral hemisphere</th>
      <th>Crus I, paravermis</th>
      <th>Crus II, lateral hemisphere</th>
      <th>Crus II, paravermis</th>
      <th>Edinger-Westphal nucleus</th>
      <th>Heschl's gyrus</th>
      <th>...</th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
      <th>transverse gyri</th>
      <th>trochlear nucleus</th>
      <th>tuberomammillary nucleus</th>
      <th>ventral tegmental area</th>
      <th>ventromedial hypothalamic nucleus</th>
      <th>vestibular nuclei</th>
      <th>zona incerta</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>0.856487</td>
      <td>-1.773695</td>
      <td>-0.678679</td>
      <td>-0.986914</td>
      <td>0.826986</td>
      <td>0.948039</td>
      <td>0.935427</td>
      <td>1.120774</td>
      <td>-1.018554</td>
      <td>0.170282</td>
      <td>...</td>
      <td>0.277830</td>
      <td>0.514923</td>
      <td>0.733368</td>
      <td>-0.104286</td>
      <td>-0.910245</td>
      <td>1.039610</td>
      <td>-0.155167</td>
      <td>-0.444398</td>
      <td>-0.901361</td>
      <td>-0.236790</td>
    </tr>
    <tr>
      <th>A1BG-AS1</th>
      <td>0.257664</td>
      <td>-1.373085</td>
      <td>-0.619923</td>
      <td>-0.636275</td>
      <td>0.362799</td>
      <td>0.353296</td>
      <td>0.422766</td>
      <td>0.346853</td>
      <td>-0.812015</td>
      <td>0.903358</td>
      <td>...</td>
      <td>1.074116</td>
      <td>0.821031</td>
      <td>1.219272</td>
      <td>0.901213</td>
      <td>-1.522431</td>
      <td>0.598719</td>
      <td>-1.709745</td>
      <td>-0.054156</td>
      <td>-1.695843</td>
      <td>-1.155961</td>
    </tr>
    <tr>
      <th>A1CF</th>
      <td>-0.089614</td>
      <td>-0.546903</td>
      <td>0.282914</td>
      <td>-0.528926</td>
      <td>0.507916</td>
      <td>0.577696</td>
      <td>0.647671</td>
      <td>0.306824</td>
      <td>0.089958</td>
      <td>0.149820</td>
      <td>...</td>
      <td>-0.030265</td>
      <td>-0.187367</td>
      <td>-0.428358</td>
      <td>-0.465863</td>
      <td>-0.136936</td>
      <td>1.229487</td>
      <td>-0.110680</td>
      <td>-0.118175</td>
      <td>-0.139776</td>
      <td>0.123829</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>0.552415</td>
      <td>-0.635485</td>
      <td>-0.954995</td>
      <td>-0.259745</td>
      <td>-1.687391</td>
      <td>-1.756847</td>
      <td>-1.640242</td>
      <td>-1.733110</td>
      <td>-0.091695</td>
      <td>0.003428</td>
      <td>...</td>
      <td>-0.058505</td>
      <td>0.207109</td>
      <td>-0.161808</td>
      <td>0.183630</td>
      <td>0.948098</td>
      <td>-0.977692</td>
      <td>0.911896</td>
      <td>-0.499357</td>
      <td>1.469386</td>
      <td>0.557998</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>0.758031</td>
      <td>1.549857</td>
      <td>1.262225</td>
      <td>1.338780</td>
      <td>-0.289888</td>
      <td>-0.407026</td>
      <td>-0.358798</td>
      <td>-0.589988</td>
      <td>0.944684</td>
      <td>-0.466327</td>
      <td>...</td>
      <td>-0.472908</td>
      <td>-0.598317</td>
      <td>-0.247797</td>
      <td>-0.282673</td>
      <td>1.396365</td>
      <td>0.945043</td>
      <td>0.158202</td>
      <td>0.572771</td>
      <td>0.073088</td>
      <td>-0.886780</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 232 columns</p>
</div></div></div>
</div>
<p>It would help to know what information is in our dataset. In other words, what is across the columns at the top? We can get a list by accessing the <code class="docutils literal notranslate"><span class="pre">columns</span></code> attribute.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Access the columns of our dataframe </span>
<span class="n">gene_df_columns</span> <span class="o">=</span> <span class="n">gene_df</span><span class="o">.</span><span class="n">columns</span> 
<span class="n">gene_df_columns</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;CA1 field&#39;, &#39;CA2 field&#39;, &#39;CA3 field&#39;, &#39;CA4 field&#39;,
       &#39;Crus I, lateral hemisphere&#39;, &#39;Crus I, paravermis&#39;,
       &#39;Crus II, lateral hemisphere&#39;, &#39;Crus II, paravermis&#39;,
       &#39;Edinger-Westphal nucleus&#39;, &#39;Heschl&#39;s gyrus&#39;,
       ...
       &#39;temporal pole, inferior aspect&#39;, &#39;temporal pole, medial aspect&#39;,
       &#39;temporal pole, superior aspect&#39;, &#39;transverse gyri&#39;,
       &#39;trochlear nucleus&#39;, &#39;tuberomammillary nucleus&#39;,
       &#39;ventral tegmental area&#39;, &#39;ventromedial hypothalamic nucleus&#39;,
       &#39;vestibular nuclei&#39;, &#39;zona incerta&#39;],
      dtype=&#39;object&#39;, length=232)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="indexing-dataframes">
<h2>Indexing Dataframes<a class="headerlink" href="#indexing-dataframes" title="Link to this heading">#</a></h2>
<p>Indexing in Pandas works slightly different than in NumPy. Similar to a dictionary, we can index dataframes by their names.</p>
<p>The syntax for indexing single locations in a dataframe is <code class="docutils literal notranslate"><span class="pre">dataframe.loc[row_label,column_label]</span></code>. To index an individual column, we use the shorthand syntax <code class="docutils literal notranslate"><span class="pre">dataframe.[column_label]</span></code>. To index an individual row, we use the syntax <code class="docutils literal notranslate"><span class="pre">dataframe.loc[row_label]</span></code>. To index by index #, we use the syntax <code class="docutils literal notranslate"><span class="pre">dataframe.iloc[index_number]</span></code>. Below are some examples on how to access rows, columns, and single values in our dataframe. For more information on indexing dataframes, visit the <a href = "https://pandas.pydata.org/docs/user_guide/indexing.html#indexing"> “Indexing and selecting data”</a> section in the Pandas User Guide.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select a single column</span>
<span class="n">column</span> <span class="o">=</span> <span class="s1">&#39;CA1 field&#39;</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gene expression values in CA1 field:&#39;</span><span class="p">)</span>
<span class="n">gene_df</span><span class="p">[</span><span class="n">column</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene expression values in CA1 field:
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>gene_symbol
A1BG        0.856487
A1BG-AS1    0.257664
A1CF       -0.089614
A2M         0.552415
A2ML1       0.758031
              ...   
ZYG11A     -0.496398
ZYG11B     -0.856866
ZYX        -1.941816
ZZEF1      -0.015748
ZZZ3       -0.924901
Name: CA1 field, Length: 20869, dtype: float64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select a single row</span>
<span class="n">row</span> <span class="o">=</span> <span class="s1">&#39;A1BG&#39;</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gene expression of &#39;</span><span class="p">,</span> <span class="n">row</span><span class="p">,</span> <span class="s1">&#39; across brain regions:&#39;</span><span class="p">)</span>
<span class="n">gene_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">row</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene expression of  A1BG  across brain regions:
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CA1 field                            0.856487
CA2 field                           -1.773695
CA3 field                           -0.678679
CA4 field                           -0.986914
Crus I, lateral hemisphere           0.826986
                                       ...   
tuberomammillary nucleus             1.039610
ventral tegmental area              -0.155167
ventromedial hypothalamic nucleus   -0.444398
vestibular nuclei                   -0.901361
zona incerta                        -0.236790
Name: A1BG, Length: 232, dtype: float64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select an individual value </span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gene expression of A1BG in CA1 field:&#39;</span><span class="p">)</span>
<span class="n">gene_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">row</span><span class="p">,</span> <span class="n">column</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene expression of A1BG in CA1 field:
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.8564873784944677
</pre></div>
</div>
</div>
</div>
<p>To select multiple different columns, you can use a <code class="docutils literal notranslate"><span class="pre">list</span></code> of all your columns of interest as so:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select multiple columns</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gene expression values in multiple regions :&#39;</span><span class="p">)</span>
<span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CA1 field&#39;</span><span class="p">,</span> <span class="s1">&#39;CA3 field&#39;</span><span class="p">,</span> <span class="s1">&#39;Crus I, lateral hemisphere&#39;</span><span class="p">]</span>
<span class="n">gene_df</span><span class="p">[</span><span class="n">columns</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene expression values in multiple regions :
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CA1 field</th>
      <th>CA3 field</th>
      <th>Crus I, lateral hemisphere</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>0.856487</td>
      <td>-0.678679</td>
      <td>0.826986</td>
    </tr>
    <tr>
      <th>A1BG-AS1</th>
      <td>0.257664</td>
      <td>-0.619923</td>
      <td>0.362799</td>
    </tr>
    <tr>
      <th>A1CF</th>
      <td>-0.089614</td>
      <td>0.282914</td>
      <td>0.507916</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>0.552415</td>
      <td>-0.954995</td>
      <td>-1.687391</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>0.758031</td>
      <td>1.262225</td>
      <td>-0.289888</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>ZYG11A</th>
      <td>-0.496398</td>
      <td>0.325555</td>
      <td>0.158885</td>
    </tr>
    <tr>
      <th>ZYG11B</th>
      <td>-0.856866</td>
      <td>0.701878</td>
      <td>0.337138</td>
    </tr>
    <tr>
      <th>ZYX</th>
      <td>-1.941816</td>
      <td>-0.681255</td>
      <td>0.872683</td>
    </tr>
    <tr>
      <th>ZZEF1</th>
      <td>-0.015748</td>
      <td>0.743609</td>
      <td>1.108376</td>
    </tr>
    <tr>
      <th>ZZZ3</th>
      <td>-0.924901</td>
      <td>0.108320</td>
      <td>-1.591413</td>
    </tr>
  </tbody>
</table>
<p>20869 rows × 3 columns</p>
</div></div></div>
</div>
</section>
<section id="subsetting">
<h2>Subsetting<a class="headerlink" href="#subsetting" title="Link to this heading">#</a></h2>
<p>Like NumPy arrays, we can subset our original dataframe to only include data that meets our criteria. Our dataframe has data on multiple different brain areas with many gene expression values. You can filter this dataframe using the following syntax:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">new_df</span> <span class="o">=</span> <span class="n">original_df</span><span class="p">[</span><span class="n">original_df</span><span class="p">[</span><span class="s1">&#39;Column of Interest&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Desired Value&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>In plain english, what this is saying is: save a dataframe from the original dataframe, where the original dataframe values in my Column of Interest are equal to my Desired Value. For more information on subsetting,  visit the <a href = "https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html"> “How do I select a subset of a DataFrame”</a> section in the Pandas documentation.</p>
<p>Below we will demonstrate how to execute this by taking a look at the <code class="docutils literal notranslate"><span class="pre">CA1</span> <span class="pre">field</span></code> column in <code class="docutils literal notranslate"><span class="pre">gene_df</span></code>. We will create a dataframe from <code class="docutils literal notranslate"><span class="pre">gene_df</span></code> that only contains genes that showed a certain level of gene expression.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a dataframe with only genes that have an expression </span>
<span class="c1"># value greater than 1.7 in &#39;CA1 field&#39; </span>
<span class="n">desired_column</span> <span class="o">=</span> <span class="s1">&#39;CA1 field&#39;</span>
<span class="n">desired_value</span> <span class="o">=</span> <span class="mf">1.7</span>
<span class="n">new_gene_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[</span><span class="n">gene_df</span><span class="p">[</span><span class="n">desired_column</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">desired_value</span><span class="p">]</span>
<span class="n">new_gene_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
      <th>Crus I, lateral hemisphere</th>
      <th>Crus I, paravermis</th>
      <th>Crus II, lateral hemisphere</th>
      <th>Crus II, paravermis</th>
      <th>Edinger-Westphal nucleus</th>
      <th>Heschl's gyrus</th>
      <th>...</th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
      <th>transverse gyri</th>
      <th>trochlear nucleus</th>
      <th>tuberomammillary nucleus</th>
      <th>ventral tegmental area</th>
      <th>ventromedial hypothalamic nucleus</th>
      <th>vestibular nuclei</th>
      <th>zona incerta</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ABCC12</th>
      <td>2.089999</td>
      <td>0.684837</td>
      <td>0.097313</td>
      <td>-0.051411</td>
      <td>-1.078900</td>
      <td>-0.912071</td>
      <td>-1.131497</td>
      <td>-0.799075</td>
      <td>-0.009423</td>
      <td>0.889132</td>
      <td>...</td>
      <td>1.289340</td>
      <td>0.885886</td>
      <td>1.271053</td>
      <td>0.650801</td>
      <td>-0.083413</td>
      <td>-0.793237</td>
      <td>-0.499512</td>
      <td>-0.762330</td>
      <td>-0.902496</td>
      <td>-0.904421</td>
    </tr>
    <tr>
      <th>ABHD17C</th>
      <td>1.716973</td>
      <td>0.601041</td>
      <td>1.132500</td>
      <td>1.354679</td>
      <td>-0.923195</td>
      <td>-0.887576</td>
      <td>-1.122027</td>
      <td>-0.926876</td>
      <td>0.339678</td>
      <td>0.208174</td>
      <td>...</td>
      <td>0.530983</td>
      <td>0.809101</td>
      <td>0.763170</td>
      <td>-0.093598</td>
      <td>-0.313468</td>
      <td>-0.164013</td>
      <td>-0.537971</td>
      <td>1.169105</td>
      <td>-0.663245</td>
      <td>-0.377181</td>
    </tr>
    <tr>
      <th>ABI1</th>
      <td>2.051762</td>
      <td>2.571777</td>
      <td>2.472188</td>
      <td>2.261170</td>
      <td>0.366138</td>
      <td>0.507783</td>
      <td>0.449606</td>
      <td>0.498424</td>
      <td>-1.365242</td>
      <td>0.472104</td>
      <td>...</td>
      <td>0.868471</td>
      <td>1.197022</td>
      <td>0.914783</td>
      <td>0.473846</td>
      <td>-2.230209</td>
      <td>-0.330684</td>
      <td>-1.153189</td>
      <td>-0.073650</td>
      <td>-1.681602</td>
      <td>-1.049258</td>
    </tr>
    <tr>
      <th>ACTB</th>
      <td>2.489711</td>
      <td>2.806688</td>
      <td>2.461655</td>
      <td>2.340131</td>
      <td>-1.296731</td>
      <td>-1.334696</td>
      <td>-1.158460</td>
      <td>-1.461027</td>
      <td>0.088209</td>
      <td>0.027236</td>
      <td>...</td>
      <td>-0.370215</td>
      <td>-0.946920</td>
      <td>-0.197363</td>
      <td>-0.094468</td>
      <td>0.926183</td>
      <td>-0.034337</td>
      <td>0.803164</td>
      <td>-0.389238</td>
      <td>0.596916</td>
      <td>0.110921</td>
    </tr>
    <tr>
      <th>ACTR2</th>
      <td>2.655049</td>
      <td>2.384445</td>
      <td>1.728348</td>
      <td>1.413585</td>
      <td>-1.377474</td>
      <td>-1.207922</td>
      <td>-1.496018</td>
      <td>-1.359462</td>
      <td>0.632764</td>
      <td>0.259402</td>
      <td>...</td>
      <td>0.215683</td>
      <td>0.142880</td>
      <td>0.080649</td>
      <td>0.312218</td>
      <td>1.793563</td>
      <td>-0.028263</td>
      <td>-0.189930</td>
      <td>-0.389064</td>
      <td>-0.484580</td>
      <td>-1.478232</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 232 columns</p>
</div></div></div>
</div>
</section>
<section id="dataframe-methods">
<h2>DataFrame Methods<a class="headerlink" href="#dataframe-methods" title="Link to this heading">#</a></h2>
<p>Pandas has many useful methods that you can use on your data, including <code class="docutils literal notranslate"><span class="pre">describe</span></code>, <code class="docutils literal notranslate"><span class="pre">mean</span></code>, and more. To learn more about all the different methods that can be used to manipulate and analyze dataframes, please visit the <a href = "https://pandas.pydata.org/docs/user_guide/index.html"> Pandas User Guide </a>. We will demonstrate some of these methods below.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">describe</span></code> method returns descriptive statistics of all the columns in our dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_df</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
      <th>Crus I, lateral hemisphere</th>
      <th>Crus I, paravermis</th>
      <th>Crus II, lateral hemisphere</th>
      <th>Crus II, paravermis</th>
      <th>Edinger-Westphal nucleus</th>
      <th>Heschl's gyrus</th>
      <th>...</th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
      <th>transverse gyri</th>
      <th>trochlear nucleus</th>
      <th>tuberomammillary nucleus</th>
      <th>ventral tegmental area</th>
      <th>ventromedial hypothalamic nucleus</th>
      <th>vestibular nuclei</th>
      <th>zona incerta</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>...</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
      <td>20869.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>0.003664</td>
      <td>0.017002</td>
      <td>0.015315</td>
      <td>-0.016633</td>
      <td>0.093686</td>
      <td>0.088810</td>
      <td>0.096118</td>
      <td>0.087608</td>
      <td>0.047124</td>
      <td>-0.042517</td>
      <td>...</td>
      <td>-0.056050</td>
      <td>-0.051731</td>
      <td>-0.049905</td>
      <td>-0.039528</td>
      <td>0.059726</td>
      <td>0.014856</td>
      <td>0.009535</td>
      <td>-0.002853</td>
      <td>-0.002527</td>
      <td>0.013018</td>
    </tr>
    <tr>
      <th>std</th>
      <td>0.924456</td>
      <td>1.129368</td>
      <td>1.078987</td>
      <td>0.897192</td>
      <td>1.146146</td>
      <td>1.118501</td>
      <td>1.172986</td>
      <td>1.136823</td>
      <td>0.973224</td>
      <td>0.500526</td>
      <td>...</td>
      <td>0.567950</td>
      <td>0.651098</td>
      <td>0.636729</td>
      <td>0.494012</td>
      <td>1.158916</td>
      <td>0.897387</td>
      <td>0.686320</td>
      <td>0.830602</td>
      <td>0.723977</td>
      <td>0.725770</td>
    </tr>
    <tr>
      <th>min</th>
      <td>-4.076424</td>
      <td>-5.923691</td>
      <td>-5.994731</td>
      <td>-3.971984</td>
      <td>-2.739924</td>
      <td>-2.662897</td>
      <td>-2.908676</td>
      <td>-2.864308</td>
      <td>-3.671242</td>
      <td>-1.666268</td>
      <td>...</td>
      <td>-1.840486</td>
      <td>-2.433961</td>
      <td>-2.412614</td>
      <td>-1.655962</td>
      <td>-6.330275</td>
      <td>-3.141490</td>
      <td>-1.977225</td>
      <td>-3.541112</td>
      <td>-2.369304</td>
      <td>-2.348784</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>-0.570475</td>
      <td>-0.644093</td>
      <td>-0.631248</td>
      <td>-0.573605</td>
      <td>-0.802651</td>
      <td>-0.778933</td>
      <td>-0.824404</td>
      <td>-0.794683</td>
      <td>-0.651691</td>
      <td>-0.414882</td>
      <td>...</td>
      <td>-0.472481</td>
      <td>-0.517562</td>
      <td>-0.529458</td>
      <td>-0.404416</td>
      <td>-0.751329</td>
      <td>-0.602619</td>
      <td>-0.497988</td>
      <td>-0.542164</td>
      <td>-0.532177</td>
      <td>-0.493921</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>-0.025821</td>
      <td>0.011189</td>
      <td>0.006140</td>
      <td>-0.044566</td>
      <td>0.100558</td>
      <td>0.098665</td>
      <td>0.109358</td>
      <td>0.096585</td>
      <td>0.022760</td>
      <td>-0.057495</td>
      <td>...</td>
      <td>-0.091029</td>
      <td>-0.083971</td>
      <td>-0.089572</td>
      <td>-0.051758</td>
      <td>0.037410</td>
      <td>-0.030474</td>
      <td>-0.023799</td>
      <td>-0.023471</td>
      <td>-0.045003</td>
      <td>-0.031623</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>0.561571</td>
      <td>0.706946</td>
      <td>0.706398</td>
      <td>0.547903</td>
      <td>0.985159</td>
      <td>0.951919</td>
      <td>1.007791</td>
      <td>0.960813</td>
      <td>0.736614</td>
      <td>0.312739</td>
      <td>...</td>
      <td>0.352796</td>
      <td>0.412321</td>
      <td>0.403382</td>
      <td>0.310551</td>
      <td>0.844723</td>
      <td>0.587382</td>
      <td>0.503613</td>
      <td>0.508410</td>
      <td>0.501191</td>
      <td>0.492262</td>
    </tr>
    <tr>
      <th>max</th>
      <td>7.062717</td>
      <td>7.387742</td>
      <td>6.413603</td>
      <td>7.178692</td>
      <td>2.679149</td>
      <td>2.717237</td>
      <td>2.963899</td>
      <td>2.857205</td>
      <td>7.552440</td>
      <td>2.234669</td>
      <td>...</td>
      <td>2.199291</td>
      <td>2.631498</td>
      <td>3.065735</td>
      <td>2.238555</td>
      <td>6.892682</td>
      <td>5.968364</td>
      <td>7.267837</td>
      <td>6.650673</td>
      <td>2.723777</td>
      <td>2.845665</td>
    </tr>
  </tbody>
</table>
<p>8 rows × 232 columns</p>
</div></div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">mean</span></code> and <code class="docutils literal notranslate"><span class="pre">std</span></code> method return the mean and standard deviation of each column in the dataframe, respectively.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_df</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CA1 field                            0.003664
CA2 field                            0.017002
CA3 field                            0.015315
CA4 field                           -0.016633
Crus I, lateral hemisphere           0.093686
                                       ...   
tuberomammillary nucleus             0.014856
ventral tegmental area               0.009535
ventromedial hypothalamic nucleus   -0.002853
vestibular nuclei                   -0.002527
zona incerta                         0.013018
Length: 232, dtype: float64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_df</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CA1 field                            0.924456
CA2 field                            1.129368
CA3 field                            1.078987
CA4 field                            0.897192
Crus I, lateral hemisphere           1.146146
                                       ...   
tuberomammillary nucleus             0.897387
ventral tegmental area               0.686320
ventromedial hypothalamic nucleus    0.830602
vestibular nuclei                    0.723977
zona incerta                         0.725770
Length: 232, dtype: float64
</pre></div>
</div>
</div>
</div>
<p>Let’s say we have two different dataframes and we would like to combine the two into one single dataframe. We can use either the <code class="docutils literal notranslate"><span class="pre">merge</span></code> or <code class="docutils literal notranslate"><span class="pre">join</span></code> Pandas methods in order to pull all of this data into one dataframe.</p>
<p><img alt="" src="http://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png" /></p>
<p>There are different types of joins/merges you can do in Pandas, illustrated <a href="http://www.datasciencemadesimple.com/join-merge-data-frames-pandas-python/">above</a>. Here, we want to do an <strong>inner</strong> merge, where we’re only keeping entries with indices that are in both dataframes. We could do this merge based on columns, alternatively.</p>
<p><strong>Inner</strong> is the default kind of join, so we do not need to specify it. And by default, join will use the ‘left’ dataframe, in other words, the dataframe that is executing the <code class="docutils literal notranslate"><span class="pre">join</span></code> method.</p>
<p>If you need more information, look at the <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html">join</a> and <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html">merge</a> documentation: you can use either of these to unite your dataframes, though join will be simpler!</p>
<p>Below is an example of how to join two separate dataframe into one, unified dataframe. We start with one dataframe with only entries from the <em>temporal pole</em> and another dataframe with only entries from the CA fields of the hippocampus. We can then join the two dataframes together using the syntax <code class="docutils literal notranslate"><span class="pre">unified_df</span> <span class="pre">=</span> <span class="pre">df_1.join(df_2)</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dataframe w/ only Temporal Pole entries </span>
<span class="n">temporal_pole_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[[</span><span class="s1">&#39;temporal pole, inferior aspect&#39;</span><span class="p">,</span> 
                            <span class="s1">&#39;temporal pole, medial aspect&#39;</span><span class="p">,</span> 
                            <span class="s1">&#39;temporal pole, superior aspect&#39;</span><span class="p">]]</span>
<span class="n">temporal_pole_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>0.277830</td>
      <td>0.514923</td>
      <td>0.733368</td>
    </tr>
    <tr>
      <th>A1BG-AS1</th>
      <td>1.074116</td>
      <td>0.821031</td>
      <td>1.219272</td>
    </tr>
    <tr>
      <th>A1CF</th>
      <td>-0.030265</td>
      <td>-0.187367</td>
      <td>-0.428358</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>-0.058505</td>
      <td>0.207109</td>
      <td>-0.161808</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>-0.472908</td>
      <td>-0.598317</td>
      <td>-0.247797</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dataframe w/ only CA field entries </span>
<span class="n">CA_field_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[[</span><span class="s1">&#39;CA1 field&#39;</span><span class="p">,</span> 
                       <span class="s1">&#39;CA2 field&#39;</span><span class="p">,</span> 
                       <span class="s1">&#39;CA3 field&#39;</span><span class="p">,</span> 
                       <span class="s1">&#39;CA4 field&#39;</span><span class="p">]]</span>
<span class="n">CA_field_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>0.856487</td>
      <td>-1.773695</td>
      <td>-0.678679</td>
      <td>-0.986914</td>
    </tr>
    <tr>
      <th>A1BG-AS1</th>
      <td>0.257664</td>
      <td>-1.373085</td>
      <td>-0.619923</td>
      <td>-0.636275</td>
    </tr>
    <tr>
      <th>A1CF</th>
      <td>-0.089614</td>
      <td>-0.546903</td>
      <td>0.282914</td>
      <td>-0.528926</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>0.552415</td>
      <td>-0.635485</td>
      <td>-0.954995</td>
      <td>-0.259745</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>0.758031</td>
      <td>1.549857</td>
      <td>1.262225</td>
      <td>1.338780</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Join the two dataframes</span>
<span class="n">df_1</span> <span class="o">=</span> <span class="n">temporal_pole_df</span>
<span class="n">df_2</span> <span class="o">=</span> <span class="n">CA_field_df</span>

<span class="n">unified_df</span> <span class="o">=</span> <span class="n">df_1</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">df_2</span><span class="p">)</span>
<span class="n">unified_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>temporal pole, inferior aspect</th>
      <th>temporal pole, medial aspect</th>
      <th>temporal pole, superior aspect</th>
      <th>CA1 field</th>
      <th>CA2 field</th>
      <th>CA3 field</th>
      <th>CA4 field</th>
    </tr>
    <tr>
      <th>gene_symbol</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>0.277830</td>
      <td>0.514923</td>
      <td>0.733368</td>
      <td>0.856487</td>
      <td>-1.773695</td>
      <td>-0.678679</td>
      <td>-0.986914</td>
    </tr>
    <tr>
      <th>A1BG-AS1</th>
      <td>1.074116</td>
      <td>0.821031</td>
      <td>1.219272</td>
      <td>0.257664</td>
      <td>-1.373085</td>
      <td>-0.619923</td>
      <td>-0.636275</td>
    </tr>
    <tr>
      <th>A1CF</th>
      <td>-0.030265</td>
      <td>-0.187367</td>
      <td>-0.428358</td>
      <td>-0.089614</td>
      <td>-0.546903</td>
      <td>0.282914</td>
      <td>-0.528926</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>-0.058505</td>
      <td>0.207109</td>
      <td>-0.161808</td>
      <td>0.552415</td>
      <td>-0.635485</td>
      <td>-0.954995</td>
      <td>-0.259745</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>-0.472908</td>
      <td>-0.598317</td>
      <td>-0.247797</td>
      <td>0.758031</td>
      <td>1.549857</td>
      <td>1.262225</td>
      <td>1.338780</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Those are the basics of working with Pandas dataframes! Circle back to this page or the resources linked within if you ever need a refresher. Next, we’ll talk about the power of SciPy for scientific analysis in Python.</p>
</section>
<section id="additional-resources">
<h2>Additional resources<a class="headerlink" href="#additional-resources" title="Link to this heading">#</a></h2>
<p>See the <a class="reference external" href="https://jakevdp.github.io/PythonDataScienceHandbook/03.00-introduction-to-pandas.html">Python Data Science Handbook</a> for a more in depth exploration of Pandas, and of course, the <a class="reference external" href="https://pandas.pydata.org/docs/user_guide/index.html">Pandas documentation</a>.</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./Chapter03"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="NumPy.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">NumPy</p>
      </div>
    </a>
    <a class="right-next"
       href="SciPy.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">SciPy</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#create-and-manipulate-dataframes">Create and Manipulate Dataframes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#dataframe-example">DataFrame example</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#indexing-dataframes">Indexing Dataframes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#subsetting">Subsetting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dataframe-methods">DataFrame Methods</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#additional-resources">Additional resources</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Ashley Juavinett & Bradley Voytek
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>