
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>3.4. Identifying Cells with Optogenetics &#8212; Teaching &amp; Learning with NWB Datasets</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet" />
  <link href="../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.acff12b8f9c144ce68a297486a2fa670.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="3.5. Chapter 3 Problem Set" href="Chapter_3_ProblemSet.html" />
    <link rel="prev" title="3.3. Aligning Spike Trains with Visual Stimuli and Behavior" href="Aligning_Spike_Trains_with_Stimuli_and_Behavior.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      <img src="../_static/nwb_learning.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Teaching & Learning with NWB Datasets</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../landing_page.html">
   Teaching and Learning with NWB Datasets
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../Chapter_01/Data_Science_In_Python.html">
   1. Data Science in Python
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_01/NumPy.html">
     1.3. NumPy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_01/Pandas.html">
     1.4. Pandas
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_01/SciPy.html">
     1.5. SciPy
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../Chapter_02/Obtaining_and_Working_with_NWB_Datasets.html">
   2. Obtaining and Working with NWB Datasets
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_02/Obtaining_Datasets_with_DANDI.html">
     2.1. Obtaining Datasets with DANDI
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_02/Working_with_NWB_format_in_Python.html">
     2.2. Working with NWB in Python
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_02/Chapter_2_ProblemSet.html">
     2.3. Chapter 2 Problem Set
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="Large_Scale_Electrophysiology.html">
   3. Large Scale Electrophysiology
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="Visualizing_Large_Scale_Recordings.html">
     3.1. Visualizing Large Scale Recordings
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="Obtaining_Raw_Electrophysiology_Data.html">
     3.2. Obtaining  Electrophysiology Data from the AllenSDK
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="Aligning_Spike_Trains_with_Stimuli_and_Behavior.html">
     3.3. Aligning Spike Trains with Visual Stimuli and Behavior
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     3.4. Identifying Cells with Optogenetics
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="Chapter_3_ProblemSet.html">
     3.5. Chapter 3 Problem Set
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../Chapter_04/Single-Cell_Electrophysiology.html">
   4. Single-Cell Electrophysiology
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_04/Obtaining_Data_For_Single_Cells.html">
     4.1. Obtaining Data for Single Cells
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_04/Analyzing_Computed_Features.html">
     4.2. Analyzing Computed Features
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../Chapter_05/Two_Photon_Imaging.html">
   5. Two Photon Imaging
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_05/Excitatory_Inhibitory_Decisions.html">
     5.1. Decision Making in Excitatory &amp; Inhibitory Cells
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_05/Retrieving_Brain_Observatory_Data.html">
     5.2. Retrieving Calcium Imaging Data from the AllenSDK
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_05/Analyzing_Two_Photon_Data.html">
     5.3. Analyzing Two Photon Data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../Chapter_05/Correlations_in_Two_Photon_Data.html">
     5.4. Correlations in Two Photon Data
    </a>
   </li>
  </ul>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/Chapter_03/Identifying_Cells_with_Optogenetics.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/NeuralDataScience/NeuralDataScience.github.io/master?urlpath=tree/Chapter_03/Identifying_Cells_with_Optogenetics.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/NeuralDataScience/NeuralDataScience.github.io/blob/master/Chapter_03/Identifying_Cells_with_Optogenetics.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#accessing-optogenetic-data">
   3.4.1. Accessing Optogenetic Data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#aligning-spikes-to-pulses">
   3.4.2. Aligning Spikes to Pulses
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#identifying-fast-paced-waveforms">
   3.4.3. Identifying Fast-Paced Waveforms
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="identifying-cells-with-optogenetics">
<h1><span class="section-number">3.4. </span>Identifying Cells with Optogenetics<a class="headerlink" href="#identifying-cells-with-optogenetics" title="Permalink to this headline">¶</a></h1>
<p>In this section, you will learn how to access the optogenetic data within the AllenSDK’s Neuropixels dataset. In these experiments, light-gated ion channels were expressed in mice in a Cre-dependent manner. These ion channels can depolarize Cre+ cells when specific wavelengths of light are shined on the brain.</p>
<p>Although it can be hard to identify cell types with extracellular electrophysiology, the integration of optogenetics into these experiments is one way to help identify subsets of cells. Only units with our cre-line of interest will respond to the light allowing for subgroups of neuronal untis to be identified and analyzed across differing genotypes. This technique is called <strong>optotagging</strong>.</p>
<p><strong>Note</strong>: the presence of light artifacts can create the appearance of false positives, and false negatives (cells that are Cre+ but do not respond to light) are nearly impossible to avoid. We will go over how to deal with these cases later in the notebook.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This will ensure that the AllenSDK is installed.</span>
<span class="c1"># If not, it will install it for you.</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">allensdk</span>
    <span class="k">if</span> <span class="n">allensdk</span><span class="o">.</span><span class="n">__version__</span> <span class="o">==</span> <span class="s1">&#39;2.11.2&#39;</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;allensdk already installed.&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span> 
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;incompatible version of allensdk installed&#39;</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">ImportError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
    <span class="o">!</span>pip install allensdk
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>allensdk already installed.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import necessary packages </span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span> 
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">xarray</span> <span class="k">as</span> <span class="nn">xr</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>

<span class="c1"># Import the Neuropixels Cache</span>
<span class="kn">from</span> <span class="nn">allensdk.brain_observatory.ecephys.ecephys_project_cache</span> <span class="kn">import</span> <span class="n">EcephysProjectCache</span>
<span class="kn">from</span> <span class="nn">allensdk.brain_observatory.ecephys.ecephys_project_api</span> <span class="kn">import</span> <span class="n">EcephysProjectWarehouseApi</span>
<span class="kn">from</span> <span class="nn">allensdk.brain_observatory.ecephys.ecephys_project_api.rma_engine</span> <span class="kn">import</span> <span class="n">RmaEngine</span>

<span class="c1"># Assign where data will be stored</span>
<span class="n">manifest_path</span> <span class="o">=</span> <span class="s1">&#39;manifest.json&#39;</span> 

<span class="c1"># Create the EcephysProjectCache object</span>
<span class="n">cache</span> <span class="o">=</span> <span class="n">EcephysProjectCache</span><span class="p">(</span><span class="n">manifest</span><span class="o">=</span><span class="n">manifest_path</span><span class="p">,</span>
                            <span class="n">fetch_api</span><span class="o">=</span><span class="n">EcephysProjectWarehouseApi</span><span class="p">(</span><span class="n">RmaEngine</span><span class="p">(</span><span class="n">scheme</span><span class="o">=</span><span class="s2">&quot;http&quot;</span><span class="p">,</span><span class="n">host</span><span class="o">=</span><span class="s2">&quot;api.brain-map.org&quot;</span><span class="p">,</span><span class="n">timeout</span><span class="o">=</span> <span class="mi">50</span> <span class="o">*</span> <span class="mi">60</span><span class="p">)))</span>          

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Packages imported and cache instance created.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/ashley/anaconda3/lib/python3.7/site-packages/dask/config.py:168: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  data = yaml.load(f.read()) or {}
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/ashley/anaconda3/lib/python3.7/site-packages/distributed/config.py:20: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  defaults = yaml.load(f)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Packages imported and cache instance created.
</pre></div>
</div>
</div>
</div>
<p>Because optogenetic channels are expressed in a Cre-dependent manner, let’s first see which Cre-lines are available.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_sessions</span> <span class="o">=</span> <span class="n">cache</span><span class="o">.</span><span class="n">get_session_table</span><span class="p">()</span>
<span class="n">col</span> <span class="o">=</span> <span class="s1">&#39;full_genotype&#39;</span>
<span class="n">all_sessions</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([&#39;Sst-IRES-Cre/wt;Ai32(RCL-ChR2(H134R)_EYFP)/wt&#39;,
       &#39;Pvalb-IRES-Cre/wt;Ai32(RCL-ChR2(H134R)_EYFP)/wt&#39;, &#39;wt/wt&#39;,
       &#39;Vip-IRES-Cre/wt;Ai32(RCL-ChR2(H134R)_EYFP)/wt&#39;], dtype=object)
</pre></div>
</div>
</div>
</div>
<p><strong>About these transgenic mouse lines</strong></p>
<p>At the time we created this notebook, there were three transgenic mouse lines available: Sst-IRES-Cre, Pvalb-IRES-Cre, and Vip-IRES-Cre. Each of these Cre-lines mark specific subsets of inhibitory neurons in the brain. You can find a brief description as well as two-photon images of these transgenic lines on the<a href = 'https://observatory.brain-map.org/visualcoding/transgenic'> Allen Brain Atlas website</a>.</p>
<p>You might also notice that these lines are crossed with a mouse line called “Ai32” – that’s a mouse line that expresses a channelrhodopsin optogenetic channel (ChR2) in a cre-dependent manner. It also expresses yellow florescent protein (EYFP) so that researchers can see the cells expressing ChR2.</p>
<div class="section" id="accessing-optogenetic-data">
<h2><span class="section-number">3.4.1. </span>Accessing Optogenetic Data<a class="headerlink" href="#accessing-optogenetic-data" title="Permalink to this headline">¶</a></h2>
<p>Before we begin downloading data, it is important to mention that there were two different means of light stimulation used in these experiments. The light used to evoke the units’ optogenetic response was switched from an LED light to a laser a little more than halfway through data collection to evoke a stronger response from the units.</p>
<p>To ensure we can see clear optogenetic responses from our units, we will be working with a session that used a laser to evoke these responses. In the cell below, we will filter our dataframe down to Pvalb-IRES-Cre mice and laser sessions only, and then download a session from a Pvalb-Cre session because they fire at high rates and are the most common in the cortex.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Laser experiments began on session 789848216</span>
<span class="n">laser_sessions</span> <span class="o">=</span> <span class="n">all_sessions</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">all_sessions</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span> <span class="o">&gt;=</span> <span class="mi">789848216</span><span class="p">]</span>

<span class="c1"># Filter for Pvalb mice</span>
<span class="n">pvalb_laser_sessions</span> <span class="o">=</span> <span class="n">laser_sessions</span><span class="p">[</span><span class="n">laser_sessions</span><span class="o">.</span><span class="n">full_genotype</span> <span class="o">==</span> <span class="s1">&#39;Pvalb-IRES-Cre/wt;Ai32(RCL-ChR2(H134R)_EYFP)/wt&#39;</span><span class="p">]</span>

<span class="n">pvalb_laser_sessions</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>published_at</th>
      <th>specimen_id</th>
      <th>session_type</th>
      <th>age_in_days</th>
      <th>sex</th>
      <th>full_genotype</th>
      <th>unit_count</th>
      <th>channel_count</th>
      <th>probe_count</th>
      <th>ecephys_structure_acronyms</th>
    </tr>
    <tr>
      <th>id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>797828357</th>
      <td>2019-10-03T00:00:00Z</td>
      <td>776061251</td>
      <td>brain_observatory_1.1</td>
      <td>107.0</td>
      <td>M</td>
      <td>Pvalb-IRES-Cre/wt;Ai32(RCL-ChR2(H134R)_EYFP)/wt</td>
      <td>611</td>
      <td>2232</td>
      <td>6</td>
      <td>[PPT, MB, APN, NOT, HPF, ProS, CA1, VISam, nan...</td>
    </tr>
    <tr>
      <th>829720705</th>
      <td>2019-10-03T00:00:00Z</td>
      <td>811322619</td>
      <td>functional_connectivity</td>
      <td>112.0</td>
      <td>M</td>
      <td>Pvalb-IRES-Cre/wt;Ai32(RCL-ChR2(H134R)_EYFP)/wt</td>
      <td>529</td>
      <td>1841</td>
      <td>5</td>
      <td>[SCig, SCop, SCsg, SCzo, POST, VISp, nan, CA1,...</td>
    </tr>
    <tr>
      <th>839557629</th>
      <td>2019-10-03T00:00:00Z</td>
      <td>821469666</td>
      <td>functional_connectivity</td>
      <td>115.0</td>
      <td>M</td>
      <td>Pvalb-IRES-Cre/wt;Ai32(RCL-ChR2(H134R)_EYFP)/wt</td>
      <td>450</td>
      <td>1853</td>
      <td>5</td>
      <td>[APN, NOT, MB, DG, CA1, VISam, nan, VISpm, LGd...</td>
    </tr>
    <tr>
      <th>840012044</th>
      <td>2019-10-03T00:00:00Z</td>
      <td>820866121</td>
      <td>functional_connectivity</td>
      <td>116.0</td>
      <td>M</td>
      <td>Pvalb-IRES-Cre/wt;Ai32(RCL-ChR2(H134R)_EYFP)/wt</td>
      <td>758</td>
      <td>2298</td>
      <td>6</td>
      <td>[APN, DG, CA1, VISam, nan, LP, VISpm, VISp, LG...</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Now, we can select one of these sessions to download the data. Below, we’ll choose session <code class="docutils literal notranslate"><span class="pre">829720705</span></code>. Depending on the speed of your internet connection, the next cell will take ~5-10 minutes to run. Do not interrupt the notebook while the data is downloading.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Download a Pvalb laser session</span>
<span class="n">session_id</span> <span class="o">=</span> <span class="mi">829720705</span>
<span class="n">session</span> <span class="o">=</span> <span class="n">cache</span><span class="o">.</span><span class="n">get_session_data</span><span class="p">(</span><span class="n">session_id</span><span class="p">)</span>
<span class="nb">type</span><span class="p">(</span><span class="n">session</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Session Downloaded.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Session Downloaded.
</pre></div>
</div>
</div>
</div>
<p>The optotagging stimuli are different than the stimulation presentations that were used in the previous section. To access the table containing the optogenetic stimuli data, execute <code class="docutils literal notranslate"><span class="pre">optogenetic_stimulation_epochs</span></code> on the session object. The table contains the start, stop, and duration of each light pulse as well as the pulse’s condition and level. The <code class="docutils literal notranslate"><span class="pre">level</span></code> column contains the power level of the light source defined by the peak voltage of the control signal delivered to the light source. The type of pulse used is stored in the <code class="docutils literal notranslate"><span class="pre">condition</span></code> column and each pulse is defined below.</p>
<ul class="simple">
<li><p>single square pulse: 5ms or 10ms duration</p></li>
<li><p>half period of a cosine wave: 1 second duration</p></li>
<li><p>2.5 ms pulses at 10 Hz: 1 second duration</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">opto_table</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">optogenetic_stimulation_epochs</span>
<span class="n">opto_table</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>start_time</th>
      <th>condition</th>
      <th>level</th>
      <th>stop_time</th>
      <th>stimulus_name</th>
      <th>duration</th>
    </tr>
    <tr>
      <th>id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>9208.46044</td>
      <td>a single square pulse</td>
      <td>2.0</td>
      <td>9208.46544</td>
      <td>pulse</td>
      <td>0.005</td>
    </tr>
    <tr>
      <th>1</th>
      <td>9210.64062</td>
      <td>a single square pulse</td>
      <td>1.7</td>
      <td>9210.65062</td>
      <td>pulse</td>
      <td>0.010</td>
    </tr>
    <tr>
      <th>2</th>
      <td>9212.37064</td>
      <td>2.5 ms pulses at 10 Hz</td>
      <td>1.7</td>
      <td>9213.37064</td>
      <td>fast_pulses</td>
      <td>1.000</td>
    </tr>
    <tr>
      <th>3</th>
      <td>9214.40076</td>
      <td>2.5 ms pulses at 10 Hz</td>
      <td>1.3</td>
      <td>9215.40076</td>
      <td>fast_pulses</td>
      <td>1.000</td>
    </tr>
    <tr>
      <th>4</th>
      <td>9216.55091</td>
      <td>2.5 ms pulses at 10 Hz</td>
      <td>2.0</td>
      <td>9217.55091</td>
      <td>fast_pulses</td>
      <td>1.000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</div>
<div class="section" id="aligning-spikes-to-pulses">
<h2><span class="section-number">3.4.2. </span>Aligning Spikes to Pulses<a class="headerlink" href="#aligning-spikes-to-pulses" title="Permalink to this headline">¶</a></h2>
<p>Unlike the stimulus presentations that we covered in the previous section, there is no built in function for aligning spikes to light pulses in the Allen SDK. Below, we have copied a function from the <a href = 'https://github.com/jsiegle/AllenSDK/blob/opto-tutorial/doc_template/examples_root/examples/nb/ecephys_optotagging.ipynb'> Allen SDK Optotagging Analysis tutorial</a> that does this for you.</p>
<p>The function below needs <code class="docutils literal notranslate"><span class="pre">bin</span> <span class="pre">edges</span></code>, <code class="docutils literal notranslate"><span class="pre">optogenetic</span> <span class="pre">trials</span></code>, and <code class="docutils literal notranslate"><span class="pre">units</span></code> in order to return the spike counts. For this reason, you must first specify your desired bins, optogenetic stimuli, and units of interest. We will be focusing units found in the <code class="docutils literal notranslate"><span class="pre">VISp</span></code> brain area. The Allen Institute has discovered that “10 ms pulses are the most useful stimulus for finding true light-evoked activity” so will subselect data with that stimulus.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assign optogenetic stimuli (trials)</span>
<span class="n">ss_10ms</span> <span class="o">=</span> <span class="n">opto_table</span><span class="p">[(</span><span class="n">opto_table</span><span class="p">[</span><span class="s1">&#39;duration&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.009</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">opto_table</span><span class="p">[</span><span class="s1">&#39;duration&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.02</span><span class="p">)]</span>

<span class="c1"># Assign units in brain area of interest (units)</span>
<span class="n">units_df</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">units</span>
<span class="n">VISp_df</span> <span class="o">=</span> <span class="n">units_df</span><span class="p">[</span><span class="n">units_df</span><span class="p">[</span><span class="s1">&#39;ecephys_structure_acronym&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;VISp&#39;</span><span class="p">]</span>

<span class="c1"># Assign time bins (time bins)</span>
<span class="n">time_resolution</span> <span class="o">=</span> <span class="mf">0.0005</span> <span class="c1"># 0.5 ms bins</span>
<span class="n">bin_edges</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="n">time_resolution</span><span class="p">)</span>



<span class="k">def</span> <span class="nf">optotagging_spike_counts</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">,</span> <span class="n">trials</span><span class="p">,</span> <span class="n">units</span><span class="p">):</span>
     
    <span class="n">time_resolution</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">))</span>
    
    <span class="c1"># list of lists of lists / each block is trial, each row in block is a unit</span>
    <span class="n">spike_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">trials</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">units</span><span class="p">))</span> <span class="p">)</span>

    <span class="c1"># unit_idx = &#39;index&#39; for the list of unit_ids made from units.index.values</span>
    <span class="c1"># will have index &#39;0-n&#39; instead of unit_id being the index</span>
    <span class="c1"># unit_id =  actual unit_id from units_df</span>
    <span class="c1"># looping for every unit_id that we specify interest in</span>
    <span class="k">for</span> <span class="n">unit_idx</span><span class="p">,</span> <span class="n">unit_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">units</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
        
        <span class="c1"># get spike times for ONE unit by subselecting a unit id</span>
        <span class="n">spike_times</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">spike_times</span><span class="p">[</span><span class="n">unit_id</span><span class="p">]</span>
    
        <span class="c1"># for each unit, loop for each trial of the specified stimulus</span>
        <span class="k">for</span> <span class="n">trial_idx</span><span class="p">,</span> <span class="n">trial_start</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">trials</span><span class="o">.</span><span class="n">start_time</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
            
            <span class="c1"># Determine what spike times for a unit fall with the duration of the trial </span>
            <span class="n">in_range</span> <span class="o">=</span> <span class="p">(</span><span class="n">spike_times</span> <span class="o">&gt;</span> <span class="p">(</span><span class="n">trial_start</span> <span class="o">+</span> <span class="n">bin_edges</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span> <span class="o">*</span> \
                       <span class="p">(</span><span class="n">spike_times</span> <span class="o">&lt;</span> <span class="p">(</span><span class="n">trial_start</span> <span class="o">+</span> <span class="n">bin_edges</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>

            <span class="n">binned_times</span> <span class="o">=</span> <span class="p">((</span><span class="n">spike_times</span><span class="p">[</span><span class="n">in_range</span><span class="p">]</span> <span class="o">-</span> <span class="p">(</span><span class="n">trial_start</span> <span class="o">+</span> <span class="n">bin_edges</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span> <span class="o">/</span> <span class="n">time_resolution</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">)</span>
            <span class="n">spike_matrix</span><span class="p">[</span><span class="n">trial_idx</span><span class="p">,</span> <span class="n">binned_times</span><span class="p">,</span> <span class="n">unit_idx</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="k">return</span> <span class="n">xr</span><span class="o">.</span><span class="n">DataArray</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s1">&#39;spike_counts&#39;</span><span class="p">,</span>
        <span class="n">data</span><span class="o">=</span><span class="n">spike_matrix</span><span class="p">,</span>
        <span class="n">coords</span><span class="o">=</span><span class="p">{</span>
            <span class="s1">&#39;trial_id&#39;</span><span class="p">:</span> <span class="n">trials</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> 
            <span class="s1">&#39;time_relative_to_stimulus_onset&#39;</span><span class="p">:</span> <span class="n">bin_edges</span><span class="p">,</span>
            <span class="s1">&#39;unit_id&#39;</span><span class="p">:</span> <span class="n">units</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>
        <span class="p">},</span>
        <span class="n">dims</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;trial_id&#39;</span><span class="p">,</span> <span class="s1">&#39;time_relative_to_stimulus_onset&#39;</span><span class="p">,</span> <span class="s1">&#39;unit_id&#39;</span><span class="p">]</span>
    <span class="p">)</span>

<span class="n">da</span> <span class="o">=</span> <span class="n">optotagging_spike_counts</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">,</span> <span class="n">ss_10ms</span><span class="p">,</span> <span class="n">VISp_df</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Spike times successfully aligned.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Spike times successfully aligned.
</pre></div>
</div>
</div>
</div>
<p>The function returns an <code class="docutils literal notranslate"><span class="pre">xarray.DataArray</span></code> that contains the spikes counts of each units. The dimensions for the array are the <code class="docutils literal notranslate"><span class="pre">trial_id</span></code> (an individual pulse from <code class="docutils literal notranslate"><span class="pre">opto_table</span></code>) the <code class="docutils literal notranslate"><span class="pre">time_relative_to_stimulus_onset</span></code>, and the <code class="docutils literal notranslate"><span class="pre">unit_id</span></code>.</p>
<p>Looking at the DataArray, we can see that in this session there were 75 different times that a 10ms single pulse was given and 46 different units we are focusing on based on brain area.</p>
<p>We can use the DataArray to plot a heatmap of the units’ response to the light pulses. The function below is also borrowed from the <a class="reference external" href="https://github.com/jsiegle/AllenSDK/blob/opto-tutorial/doc_template/examples_root/examples/nb/ecephys_optotagging.ipynb">Allen SDK Optogenetic Analysis tutorial</a>. The function takes our DataArray and reduces the data within by computing the mean spike counts across the <code class="docutils literal notranslate"><span class="pre">trial_id</span></code> dimension for every unit in the array. The reduced data is then plotted with respect to the time bins that we specified earlier.</p>
<p>For more information on how to use <code class="docutils literal notranslate"><span class="pre">xarray.DataArray</span></code> objects, please visit the <a href = 'http://xarray.pydata.org/en/stable/generated/xarray.DataArray.mean.html#xarray.DataArray.mean'> xarray original documentation online</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_optotagging_response</span><span class="p">(</span><span class="n">da</span><span class="p">):</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
    
    <span class="c1"># Plot mean firing rate of each unit across each trial</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">da</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="s1">&#39;trial_id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">T</span> <span class="o">/</span> <span class="n">time_resolution</span><span class="p">,</span> 
               <span class="n">extent</span><span class="o">=</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">),</span>
                       <span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">VISp_df</span><span class="p">)],</span>
               <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>    <span class="c1"># Does this have to be hard coded?</span>
    
    <span class="c1"># Plot Vertical lines indicating stimulus window </span>
    <span class="k">for</span> <span class="n">bound</span> <span class="ow">in</span> <span class="p">[</span><span class="mf">0.0005</span><span class="p">,</span> <span class="mf">0.0095</span><span class="p">]:</span> <span class="c1"># THIS IS HARD CODED; CAN WE FIX IT</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">bound</span><span class="p">,</span> <span class="n">bound</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">VISp_df</span><span class="p">)],</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Time (s)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Unit #&#39;</span><span class="p">)</span>

    <span class="n">cb</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">fraction</span><span class="o">=</span><span class="mf">0.046</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="mf">0.04</span><span class="p">)</span>
    <span class="n">cb</span><span class="o">.</span><span class="n">set_label</span><span class="p">(</span><span class="s1">&#39;Mean firing rate (Hz)&#39;</span><span class="p">)</span>
    
<span class="n">plot_optotagging_response</span><span class="p">(</span><span class="n">da</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Pvalb-cre mice session&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/Identifying_Cells_with_Optogenetics_18_0.png" src="../_images/Identifying_Cells_with_Optogenetics_18_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Removed function</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
    
<span class="c1"># Plot mean firing rate of each unit across each trial</span>
<span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">da</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="s1">&#39;trial_id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">T</span> <span class="o">/</span> <span class="n">time_resolution</span><span class="p">,</span> 
           <span class="n">extent</span><span class="o">=</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">),</span>
                   <span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">VISp_df</span><span class="p">)],</span>
           <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">)</span>    <span class="c1"># Removed hard coded color bar range</span>
    
<span class="c1"># Plot Vertical lines indicating stimulus window </span>
<span class="k">for</span> <span class="n">bound</span> <span class="ow">in</span> <span class="p">[</span><span class="mf">0.0005</span><span class="p">,</span> <span class="mf">0.0095</span><span class="p">]:</span> <span class="c1"># THIS IS HARD CODED; CAN WE FIX IT</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">bound</span><span class="p">,</span> <span class="n">bound</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">VISp_df</span><span class="p">)],</span> <span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Time (s)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Unit #&#39;</span><span class="p">)</span>

<span class="n">cb</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">fraction</span><span class="o">=</span><span class="mf">0.046</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="mf">0.04</span><span class="p">)</span>
<span class="n">cb</span><span class="o">.</span><span class="n">set_label</span><span class="p">(</span><span class="s1">&#39;Mean firing rate (Hz)&#39;</span><span class="p">)</span>
    
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Pvalb-cre mice session&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/Identifying_Cells_with_Optogenetics_19_0.png" src="../_images/Identifying_Cells_with_Optogenetics_19_0.png" />
</div>
</div>
<p>A useful analysis with this data would be to compare the baseline firing rate (i.e. firing rate before stimulus) to the evoked firing rate. We can do this by first slicing our time bins into two timeframes, <code class="docutils literal notranslate"><span class="pre">baseline</span></code> and <code class="docutils literal notranslate"><span class="pre">evoked</span></code>. The <code class="docutils literal notranslate"><span class="pre">baseline</span></code> timeframe will begin at the start of our time bins and end just before the stimulus onset. The <code class="docutils literal notranslate"><span class="pre">evoked</span></code> timeframe will begin immediately after the onset of the stimulus and end right before the end of the stimulus suration.</p>
<p>We will then reduce our data to only the mean firing rate of each unit within each time frame.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assign time before stimulus occurs</span>
<span class="n">baseline</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">time_relative_to_stimulus_onset</span><span class="o">=</span><span class="nb">slice</span><span class="p">(</span><span class="o">-</span><span class="mf">0.01</span><span class="p">,</span><span class="o">-</span><span class="mf">0.002</span><span class="p">))</span> <span class="c1"># ALSO HARD CODED</span>

<span class="c1"># Calculate Mean firing rate of each unit for time before the stimulus window </span>
<span class="n">baseline_rate</span> <span class="o">=</span> <span class="n">baseline</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="s1">&#39;time_relative_to_stimulus_onset&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="s1">&#39;trial_id&#39;</span><span class="p">)</span> <span class="o">/</span> <span class="mf">0.008</span>

<span class="c1"># Assign time within stimulus window</span>
<span class="n">evoked</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">time_relative_to_stimulus_onset</span><span class="o">=</span><span class="nb">slice</span><span class="p">(</span><span class="mf">0.001</span><span class="p">,</span><span class="mf">0.009</span><span class="p">))</span>

<span class="c1"># Calculate Mean firing rate of each unit for time within the stimulus window </span>
<span class="n">evoked_rate</span> <span class="o">=</span> <span class="n">evoked</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="s1">&#39;time_relative_to_stimulus_onset&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="s1">&#39;trial_id&#39;</span><span class="p">)</span> <span class="o">/</span> <span class="mf">0.008</span>

<span class="n">evoked_rate</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre>&lt;xarray.DataArray &#x27;spike_counts&#x27; (unit_id: 52)&gt;
array([  0.        , 198.33333333,   0.        ,   1.66666667,
         6.66666667,   0.        ,   5.        ,  15.        ,
         6.66666667,   3.33333333,   0.        ,   6.66666667,
         1.66666667, 131.66666667,  10.        ,  55.        ,
         1.66666667,   0.        ,   1.66666667, 110.        ,
         0.        ,   6.66666667,  13.33333333,  10.        ,
         0.        ,   1.66666667,  81.66666667,  18.33333333,
         1.66666667,   0.        ,   8.33333333,   1.66666667,
        11.66666667,   0.        ,  38.33333333,  11.66666667,
       235.        ,   1.66666667,  10.        ,   0.        ,
        21.66666667,   0.        ,   3.33333333,   1.66666667,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   3.33333333,   0.        ,   0.        ])
Coordinates:
  * unit_id  (unit_id) int64 951132060 951132054 ... 951132308 951132419</pre></div></div>
</div>
<p>We now have two DataArrays, one that contains the mean firing rate of the units at baseline, and one that contains the mean firing rate of the units during the pulse. Plotting a scatter plot with these two areas will show us how a unit’s basline firing rate compares to its evoked firing rate.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">baseline_rate</span><span class="p">,</span> <span class="n">evoked_rate</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="c1"># Black line though center represents no change in rate (1:1)</span>
<span class="n">axis_limit</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">axis_limit</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span><span class="n">axis_limit</span><span class="p">],</span> <span class="s1">&#39;:k&#39;</span><span class="p">)</span>

<span class="c1"># Red line represents a 2x change in rate (2:1)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">axis_limit</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span><span class="n">axis_limit</span><span class="o">*</span><span class="mi">2</span><span class="p">],</span> <span class="s1">&#39;:r&#39;</span><span class="p">)</span>

<span class="c1"># Restrict the axis for visualization</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">axis_limit</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">axis_limit</span><span class="p">])</span>

<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Baseline rate (Hz)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Evoked rate (Hz)&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/Identifying_Cells_with_Optogenetics_23_0.png" src="../_images/Identifying_Cells_with_Optogenetics_23_0.png" />
</div>
</div>
<p>Each point represent a neural unit. The black line represents a 1:1 ratio in the evoked firing rate to the baseline firing rate of that unit. Any unit below that black line had a larger response at baseline than it did to the light pulse. The red line represents a 2:1 ratio in the evoked firing rate to the baseline firing rate of that unit. We can subselect units whose evoked firing rate is 2x larger than baseline to define the cre+ units.</p>
</div>
<div class="section" id="identifying-fast-paced-waveforms">
<h2><span class="section-number">3.4.3. </span>Identifying Fast-Paced Waveforms<a class="headerlink" href="#identifying-fast-paced-waveforms" title="Permalink to this headline">¶</a></h2>
<p>Defining what units are cre+ from those that are cre- is the first step in comparing optogenetic responses. Below we will create two lists, one with IDs of units who had an evoked firing rate 2x greater than baseline, and one with IDs of units whose evoked response was smaller than 2x the baseline.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Return units whose firing rates doubled when light was on</span>
<span class="n">cre_pos_units</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">unit_id</span><span class="p">[(</span><span class="n">evoked_rate</span> <span class="o">/</span> <span class="p">(</span><span class="n">baseline_rate</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="c1"># add 1 to prevent divide-by-zero errors</span>

<span class="c1"># also get cre neg units</span>
<span class="n">cre_neg_units</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">unit_id</span><span class="p">[(</span><span class="n">evoked_rate</span> <span class="o">/</span> <span class="p">(</span><span class="n">baseline_rate</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Pvalb+ Cre units&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cre_pos_units</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39; </span><span class="se">\n</span><span class="s1"> Pvalb- Cre units&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cre_neg_units</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Pvalb+ Cre units
[951132054 951132138 951132140 951132159 951132184 951132212 951132205
 951132224 951132236]
 
 Pvalb- Cre units
[951132060 951132379 951132066 951132064 951132062 951132381 951132070
 951132078 951132090 951132092 951132112 951132124 951132136 951132393
 951132395 951132153 951132175 951132173 951132171 951132179 951132399
 951132189 951132197 951132195 951132203 951132401 951132218 951132208
 951132226 951132403 951132220 951132459 951132234 951132232 951132230
 951132405 951132254 951132258 951132288 951132290 951132302 951132308
 951132419]
</pre></div>
</div>
</div>
</div>
<p>Lets take a look at the wavefroms of cre+ units and see how they compare to cre- units.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
    
<span class="k">for</span> <span class="n">unit_id</span> <span class="ow">in</span> <span class="n">cre_neg_units</span><span class="p">:</span>
    
    <span class="n">peak_channel</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">units</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">unit_id</span><span class="p">]</span><span class="o">.</span><span class="n">peak_channel_id</span>
    <span class="n">wv</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">mean_waveforms</span><span class="p">[</span><span class="n">unit_id</span><span class="p">]</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">channel_id</span> <span class="o">=</span> <span class="n">peak_channel</span><span class="p">)</span>    
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wv</span><span class="o">.</span><span class="n">time</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">wv</span><span class="p">,</span> <span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>    

<span class="k">for</span> <span class="n">unit_id</span> <span class="ow">in</span> <span class="n">cre_pos_units</span><span class="p">:</span>
    
    <span class="n">peak_channel</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">units</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">unit_id</span><span class="p">]</span><span class="o">.</span><span class="n">peak_channel_id</span>
    <span class="n">wv</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">mean_waveforms</span><span class="p">[</span><span class="n">unit_id</span><span class="p">]</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">channel_id</span> <span class="o">=</span> <span class="n">peak_channel</span><span class="p">)</span>    
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wv</span><span class="o">.</span><span class="n">time</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">wv</span><span class="p">,</span> <span class="s1">&#39;c&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Time (ms)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Amplitude (uV)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mf">1.5</span><span class="p">])</span> <span class="c1"># Plot just 1.5 ms</span>

<span class="c1"># Create the legend</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">([</span><span class="s1">&#39;Pvalb- Waveforms&#39;</span><span class="p">,</span><span class="s1">&#39;Pvalb+ Waveforms&#39;</span><span class="p">])</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span>
<span class="n">leg</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_legend</span><span class="p">()</span>
<span class="n">leg</span><span class="o">.</span><span class="n">legendHandles</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">leg</span><span class="o">.</span><span class="n">legendHandles</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;c&#39;</span><span class="p">)</span>

<span class="c1">#plt.plot([1.0, 1.0],[-160, 100],&#39;:c&#39;) # REMOVING THIS UNTIL WE KNOW WHAT IT IS</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/Identifying_Cells_with_Optogenetics_29_0.png" src="../_images/Identifying_Cells_with_Optogenetics_29_0.png" />
</div>
</div>
<p>As you can see, our Pvalb-positive waveforms (in cyan) are more narrow than the Pvalb-negative waveforms. These are indeed different cell types, marked by their ability to spike very fast!</p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./Chapter_03"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="Aligning_Spike_Trains_with_Stimuli_and_Behavior.html" title="previous page"><span class="section-number">3.3. </span>Aligning Spike Trains with Visual Stimuli and Behavior</a>
    <a class='right-next' id="next-link" href="Chapter_3_ProblemSet.html" title="next page"><span class="section-number">3.5. </span>Chapter 3 Problem Set</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Victor Magdaleno-Garcia & Ashley Juavinett<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>